# QC exercises
```{r, fig.align = 'center',out.width= '20%', echo=FALSE }
knitr::include_graphics(path = "figures/exercise.png", auto_pdf = TRUE)
```
Now you will carry out Quality checking and control of a whole genome shotgun study and a 16S rRNA study. You will run through the Quality checking and Quality control using the examples to help you. There will be questions to answer after which there will be info on the data and what the QC has shown.

## Whole shotgun metagenomics {#shotgun}
```{r, fig.align = 'center',out.width= '20%', echo=FALSE }
knitr::include_graphics(path = "figures/kraken.png", auto_pdf = TRUE)
```
The data from this exercise comes from the following link:

https://www.ebi.ac.uk/ena/data/search?query=shotgun

This contains the genomic content of many organisms.

The read data is within the exercise_1 directory. The reads have been subsampled so you will be working on 100,000 paired reads.

Carry out `fastqc` on the raw reads and answer the below questions.

1. What may be causing the weird GC content at the start of the reads?
2. The \"Per sequence GC content\" has a weird peak. Why might this be?

Carry out `trimmomatic` quality control, deciding on the sliding window and the length threshold. After `trimmomatic` carry out `fastqc` on the trimmed and filtered reads to determine if it was successful. Afterwards answer the following questions.

1. Is the number of quality controlled reads above 60,000?
2. Does the quality for either R1 or R2 drop below Q24, if so at what position?
3. Is the weird GC content still at the start of the read?
4. Does the \"Per sequence GC content\" still have a weird peak?

In this case it helps to know the adapter sequence, that was used for the Illumina sequencing, was Nextera. Nextera is decent but I and others online have experienced it has a slight bias. It prefers to ligate to certain sequence patterns, this bias is then shown as the GC content bias at the start of reads. Fortunately the rest of the GC content is fine. 

In previous tests I have trimmed off the first 20bp to remove this bias and not removed the first 20bp. I found there was little difference in the downstream results for the two methods. 

As for the weird \"Per sequence GC content\" peak we must remember that the metagenome sample contains many different organisms with different GC contents so this would be normal.

## 16S rRNA Metagenetics {#metagenetics}
```{r, fig.align = 'center',out.width= '20%', echo=FALSE }
knitr::include_graphics(path = "figures/chime.png", auto_pdf = TRUE)
```
Before starting this exercise I must note this is not the normal method to quality control 16S rRNA data. Both QIIME2 and MOTHUR have their own methods but it can be useful to check the quality of 16S rRNA data.

With that in mind we will only use fastQC to check the quality and features of a 16S rRNA dataset.

The dataset can be found in the following link:

https://www.ebi.ac.uk/ena/data/view/SRR2081030

The dataset in the cluster is the full dataset. It can be found in the exercise_2 directory.

Once you have looked at the `fastqc` output of the reads you may be worried by its appearance, however this is normal for 16S rRNA data. Some explanations for the features are:

- The base quality of R2 is very low towards the end. For most 16S rRNA studies the paired reads overlap and so poor quality bases at the end of R1 are corrected by the overlapping R2 reads and vice versa.
- The \"Per base sequence content\" is heavily biased. This again is normal as we are targeting a specific region within the genome, a 16S rRNA region. This means a lot of the sequences will share very similar content. The fact we are targeting a specific region also makes the \"Per sequence GC content\", \"Sequence duplication levels\", and \"overrepresented sequences\" look worrying and warnings are given. These warnings are designed to help with whole genome data and so they are not useful in this circumstance.

These features show that it is important to understand what your data contains and what analyses you will be carrying out with the data.