--- 
title: "Long read Quality Control"
author: "Matthew R. Gemmell & Helen Hipperson"
date: "`r Sys.Date()`"
site: bookdown::bookdown_site
output: bookdown::gitbook
documentclass: book
bibliography: [book.bib, packages.bib]
biblio-style: apalike
link-citations: yes
favicon: figures/NEOF_favicon.png
description: NEOF book for the Long read quality control workshop
cover-image: "figures/NEOF.png"
---

```{r cite-packages, include = FALSE}
# automatically create a bib database for R packages
# add any packages you want to cite here
knitr::write_bib(c(
  .packages(), 'bookdown', 'webexercises'
), 'packages.bib')
```

```{r, echo=FALSE}
#This book uses ionicons icons using the icons package
#https://github.com/mitchelloharawild/icons
#To set this up on your rstudio use the following commands
# install.packages("remotes")
#remotes::install_github("mitchelloharawild/icons")
#download_ionicons()
```

```{r, echo=FALSE}
#Change colour, border, and text of code chunks
#Check style.css for .Rchunk
#https://stackoverflow.com/questions/65627531/change-r-chunk-background-color-in-bookdown-gitbook
#https://bookdown.org/yihui/rmarkdown-cookbook/chunk-styling.html
knitr::opts_chunk$set(class.source="Rchunk") 
```

<center>
![](figures/NEOF_rounded_corners.png){style="width:300px"}
</center>

# Overview
<center>
![](figures/well.png){style="width:200px"}
</center>

The most popular long read sequencers are produced by ONT and PacBio.
It is beneficial to understand how they work and their error profiles to fully utilise the data they produce.
This course aims to introduce you to ONT and PacBio quality control and checking.

Sessions will start with a brief presentation followed by self-paced computer practicals guided by an online interactive book. The book will contain theory, practice code, and exercises. Multiple choice questions will guide the interpretation of results.

At the end of this course learners will be able to:

- Explain how ONT & PacBio sequencing works.
- Understand the advantages and disadvantages of both technologies.
- Quality check ONT data with NanoFilt and NanoPlot.
- Remove ONT adaptors with Porechop.
- Filter ONT reads by length and quality with NanoFilt.
- Quality check & filter PacBio data with SequelTools.
- Convert PacBio BAM files to fastq.
- Generate PacBio CCS reads from subreads.
- Quality check PacBio CCS reads with RabbitQCPlus.

## Table of contents {-}

```{r, echo=FALSE, align="centre"}
#Create 2 column dataframe
column_1 <- c(
  #1 Data intro
  '[**Data intro**](#data)', 
 '[![](figures/Play.png){style="height:150px; border-radius:15px; border:5px solid #333333"}](#data)',
 #3 ONT setup
    '[**ONT setup**](#ontsetupchap)', 
 '[![](figures/conda.png){height="150"}](#ontsetupchap)',
 #5 Nanoplot
   '[**NanoPlot**](#nanoplotchap)', 
 '[![](figures/nanoplot_hexes.png){height="150"}](#nanoplotchap)',
 #7 ONT final check
   '[**ONT final check**](#ontcheckchap)', 
 '[![](figures/finish.png){style="height:150px; border-radius:15px; border:5px solid #333333; background:white"}](#ontcheckchap)',
  #9 CLR setup
   '[**CLR setup**](#clrsetupchap)', 
 '[![](figures/clr.png){style="height:100px; border-radius:15px; border:5px solid #333333"}](#clrsetupchap)',
   #11 CLR filter
   '[**CLR filtering**](#clrfiltchap)', 
 '[![](figures/quality_filtering.png){height="150"}](#clrfiltchap)',
    #13 appendix
   '[**Appendix**](#mamba_install)', 
 '[![](figures/mamba_logo.png){style="height:150px; border-radius:15px; border:5px solid #333333"}](#mamba_install)'
)
column_2 <- c(
  #2 ONT
  '[**ONT background**](#ont)', 
 '[![](figures/ONT_logo_wo_text.png){style="height:150px; border-radius:15px; border:5px solid #333333"}](#ont)',
 #4 porechop
   '[**Porechop**](#porechopchap)', 
 '[![](figures/porechop_logo_knife.png){height="100"}](#porechopchap)',
 #6 nanofilt
   '[**NanoFilt**](#nanofiltchap)', 
 '[![](figures/quality_filtering.png){height="150"}](#nanofiltchap)',
 #8 pb background
   '[**PacBio background**](#pb)', 
 '[![](figures/logo_PacBio_web_RGB-1.png){style="height:150px; border-radius:15px; border:5px solid #333333; background:white"}](#pb)',
   #10 CLR qc
   '[**CLR QC**](#clrqcchap)', 
 '[![](figures/qc_inspection.png){height="150"}](#clrqcchap)',
    #12 CCS
   '[**CCS**](#ccschap)', 
 '[![](figures/ccs.png){style="height:150px; border-radius:15px; border:5px solid #333333"}](#ccschap)',
 #14 empty
 "",""
)
table_df <- data.frame(column_1, column_2)
#Kable and kableextra libraries
library(knitr)
library(kableExtra)
#Create table
#ALign both column in centre (align = "cc")
table_df %>%
  kbl(align= "cc", col.names = NULL) %>%
  kable_styling(position = "center")
```

<a rel="license" href="http://creativecommons.org/licenses/by-nc-sa/4.0/"><img alt="Creative Commons Licence" style="border-width:0" src="https://i.creativecommons.org/l/by-nc-sa/4.0/88x31.png" /></a><br />This work is licensed under a <a rel="license" href="http://creativecommons.org/licenses/by-nc-sa/4.0/">Creative Commons Attribution-NonCommercial-ShareAlike 4.0 International License</a>.