# (PART\*) ONT {.unnumbered}

# ONT background {#ont}
<center>
![](figures/ONT_logo.png){style="width:600px; border-radius:5px; border:5px solid white; background:null"}
</center>

In this section we are going to carry out a quick QC of ONT data using the tool suite NanoPack and the tool Porechop. For this we will use fastq files.

Before carrying out analysis we will cover some background of ONT data. This will include [file formats](#ont_file_formats), & [basecalling](#ont_basecalling).

## File formats {#ont_file_formats}
<center>
![](figures/files.png){style="width:200px; border-radius:5px; border:5px solid white; background:white"}
</center>

Although we will only be using fastq files for this tutorial it is important to know the other file formats you may encounter when working with ONT data.

### POD5
<center>
![](figures/arrow-logo_chevrons_black-txt_white-bg.png){style="width:150px; border-radius:15px"}
</center>

The POD5 format is the ONT created format to store nanopore DNA data in an easily accessible manner. The files end with the extension `.pod5` and is normally created by the ONT tool `MinKNOW`. It is a collection of [Apache Arrow tables](https://arrow.apache.org/docs/java/table.html) (it is very unlikely you need to understand Apache Arrow tables).

The POD5 file format contains the following information:

- __Reads Table:__ Describes the metadata for each read.
- __Signal Table:__ Contains the signal data from the sequencer.
- __Run Info Table:__ Contains information on the `MinKNOW` runs that created the reads that are in the POD5 file.

The POD5 format is quite new and was released as an early preview in May 2022. Prior to POD5 they used Fast5 files. They changed to POD5 with the following design goals in mind:

- Good write performance for MinKNOW
- Recoverable if the writing process crashes
- Good read performance for downstream tools, including basecall model generation
- Efficient use of space
- Straightforward to implement and maintain
- Extensibility

POD5 links:

- [Main POD5 github page](https://github.com/nanoporetech/pod5-file-format)
- [POD5 File Format Design Details](https://github.com/nanoporetech/pod5-file-format/blob/master/docs/DESIGN.md)
- [POD5 Format Specification](https://github.com/nanoporetech/pod5-file-format/blob/master/docs/SPECIFICATION.md)
- [POD5 Tools commands](https://github.com/nanoporetech/pod5-file-format/blob/master/python/pod5/README.md#tools)

### Fast5
<center>
![](figures/fast5.png){style="width:150px; border-radius:5px; border:5px solid white; background:white"}
</center>

The Fast5 format was the previous format used to store raw data from ONT machine. It is now a legacy format. However, if you use older data for your research you may encounter it.

The format contains the signal data from the pore which can be processed into more useful files.

Fast5 files can contain:

+ Raw signal data
+ Run metadata
+ fastq-basecalls
+ Other additional analyses

The Fast5 file format is an implementation of the HDF5 file format specifically for ONT sequencing data. For more information on Fast5 and a tool to interface with these types of file please see:
https://github.com/nanoporetech/ont_fast5_api

You can convert a Fast5 file to a Pod5 file with the pod5 tools command [`pod5 convert fast5`](https://github.com/nanoporetech/pod5-file-format/blob/master/python/pod5/README.md#pod5-convert-fast5)

### Summary file

The MinION and GridION output a single sequencing summary file in addition to the POD5/Fast5 files. This file contains metdata which describes each sequenced read.

We will not use POD5, Fast5 or summary files for this tutorial as they are not needed most of the time. It is likely these files will not be initially provided to you by a sequencing service as they are very large. However, if you do require them you can always ask but be careful with how long the sequencing centre will retain your data.

### BAM file
<center>
![](figures/bam.png){style="width:100px; border-radius:5px; border:5px solid white; background:white"}
</center>

__BAM__ files (`.bam`) are the binary version of __SAM__ files (`.sam`).

- __SAM__ stands for "Sequence Alignment/Map" (the B in __BAM__ is Binary).
- __SAM__ files are tab delimited and contain alignment information. 
- __BAM__ files are not human readable whilst __SAM__ files are.
- __SAM__ files are larger than __BAM__.

Generally programs that can work on __SAM__ files can also work on __BAM__ files. Due to the size difference it is preferable to store data in __BAM__ format over __SAM__. Even though a BAM file is smaller than its matching __SAM__ file, __BAM__ files are still very large and can be >100GB.

It can be useful to contain unaligned reads from sequencing machines (e.g\. PacBio and ONT) in __BAM__ files as they can contain more metadata in the header and per-record auxiliary tags compared to __fastq__ files. 

If you are working with __SAM__/__BAM__ files the following link will prove useful: 
https://www.htslib.org/

For more information on the __SAM__ and __BAM__ format please see: 
https://samtools.github.io/hts-specs/SAMv1.pdf


### fastq file
<center>
![](figures/fastq.png){style="width:150px; border-radius:5px; border:5px solid white; background:white"}
</center>

The __fastq__ file format is very consistent and so there is no real difference between __fastq__ files for Illumina, PacBio, and ONT data.

All __fastq__ files will contain a number of lines divisible by 4. This is because each entry/sequence will take up four lines consisting of the following information:

1. Header for __fastq__ entry known as the __fastq header__. 
    + This always begins with a __\'@\'__
    + This is where you might see the most difference between different data. Different machines and different public databases will use different formats for __fastq__ headers.
2. Sequence content
3. Quality header 
    + Always begins with a \'+\'. Sometimes also contains the same information as the __fastq__ header.  
4. Quality
    + Each base in the 2nd line will have a corresponding quality value in this line.
    + Note this uses Phred encoding, of which there are different formats. Most, if not all, new files will use Phred+33 but be careful if you are using older data as it may use a different one. See the \"Encoding\" section in the following link for more info: https://en.wikipedia.org/wiki/FASTQ_format.
    +  NOTE: __\'@\'__ can be used as a quality value.

An example of the information of one sequence/entry is:

__\@Sequence 1__  
__CTGTTAAATACCGACTTGCGTCAGGTGCGTGAACAACTGGGCCGCTTT__  
__+__  
__=<<<=>\@\@\@ACDCBCDAC\@BAA\@BA\@BBCBBDA\@BB\@>CD\@A\@B?B\@\@__

## Basecalling {#ont_basecalling}
<center>
![](figures/guppy.png){style="width:200px; border-radius:5px; border:5px solid white; background:white"}
</center>

In short, basecalling is the conversion of ONT signals to bases and quality. This step converts the __POD5/Fast5__ files to __BAM__ and/or __fastq__ files.

There are many tools to carry out basecalling with ONT data. ONT sequencing machines carry this out with in built programs. However, if you are interested in the tools used for this the primary ones are MinKNOW, Guppy, Dorado, and Albacore. Unfortunately it is quite hard to find information about these tools unless you own an ONT machine.

We will not go into any more detail about basecalling as this is normally carried out by the machine/sequencing centre. Instead please find some useful links below if you would like to know more.

- MinKNOW
  - [Overview on working with MinKNOW](https://community.nanoporetech.com/docs/prepare/library_prep_protocols/experiment-companion-minknow/v/mke_1013_v1_revde_11apr2016)
  - [Basecalling algorithms](https://community.nanoporetech.com/technical_documents/data-analysis/v/datd_5000_v1_revt_22aug2016/basecalling-algorithms)
- Guppy
  - [Basecalling with Guppy tutorials](https://denbi-nanopore-training-course.readthedocs.io/en/latest/basecalling/index.html)
  - [Input and output files](https://community.nanoporetech.com/docs/prepare/library_prep_protocols/Guppy-protocol/v/gpb_2003_v1_revax_14dec2018/input-and-output-files)
- Dorado
  - [Github page](https://github.com/nanoporetech/dorado)
- Literature
  - [Comprehensive benchmark and architectural analysis of deep learning models for nanopore sequencing basecalling](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-023-02903-2)