# ONT
```{r, fig.align = 'center',out.width= '60%', echo=FALSE }
knitr::include_graphics(path = "figures/ONT_logo.png", auto_pdf = TRUE)
``` 

In this section we are going to carry out a quick QC of ONT data using the tool suite NanoPack and the tool Porechop. For this we will use the fastq files.

## File formats
```{r, fig.align = 'center',out.width= '15%', echo=FALSE }
knitr::include_graphics(path = "figures/files.png", auto_pdf = TRUE)
``` 

Although we will only be using fastq files for this tutorial it is important to know the other file formats you may encounter when working with ONT data.

### Fast5
```{r, fig.align = 'center',out.width= '10%', echo=FALSE }
knitr::include_graphics(path = "figures/fast5.png", auto_pdf = TRUE)
``` 

The raw data from ONT machines come as Fast5 (.fast5) files. This contains the signal data from the pore which can be processed into more useful files.

Fast5 files can contain:

+ Raw signal data
+ Run metadata
+ fastq-basecalls
+ Other additional analyses

The Fast5 file format is an implementation of the HDF5 file format specifically for ONT sequencing data. For more information on Fast5 and a tool to interface with these types of file please see:
https://github.com/nanoporetech/ont_fast5_api

### Summary file

The MinION and GridION output a single sequencing summary file in addition to the Fast5 files. This file contains metdata which descirbes each sequenced read.

We will not use Fast5 or summary files for this tutorial as they are not needed most of the time. It is likely these files will not be intially provided to you by a sequencing service as they are large and not often needed. However, if you do require them you can always ask but be careful with how long the sequencing centre will retain your data.

### BAM file
```{r, fig.align = 'center',out.width= '10%', echo=FALSE }
knitr::include_graphics(path = "figures/bam.png", auto_pdf = TRUE)
``` 

BAM files (.bam) are the binary version of SAM files (.sam). This means:

- BAM files are not human readable whilst SAM files are.
- SAM files are larger than BAM.

Generally programs that can work on SAM files can also work on BAM files. Due to the size difference it is preferable to store data in BAM format over SAM. Even though a BAM file is smaller than their matching SAM file, BAM files are still very large and can be >100GB.

SAM stands for "Sequence Alignment/Map" (the B in BAM is Binary). SAM files are tab delimited and contain alignment information. 

It can be useful to contain unaligned reads from sequencing machines (e.g. PacBio and ONT) in BAM files as they can contain more metadata in the header and per-record auxiliary tags compared to fastq files. 

If you are working with SMA/BAM files the following link will prove useful: 
https://www.htslib.org/

For more information on the SAM and BAM format please see: 
https://samtools.github.io/hts-specs/SAMv1.pdf


### fastq file
```{r, fig.align = 'center',out.width= '10%', echo=FALSE }
knitr::include_graphics(path = "figures/fastq.png", auto_pdf = TRUE)
``` 

The fastq file format is very consistent and so there is no real difference between fastq files for Illumina, PacBio, and ONT data.

All fastq files will contain a number of lines divisible by 4. This is because each entry/sequence will take up four lines consisting of the following information:

1. Header for fastq entry known as the fastq header. 
    + This always begins with a __‘@’__
    + This is where you might see the most difference between different data. Different machines and different public databases will use different formats for fastq headers.
2. Sequence content
3. Quality header 
    + Always begins with a ‘+’. Sometimes also contains the same information as fastq header.  
4. Quality
    + Each base in the 2nd line will have a corresponding quality value in this line.
    + Note this uses Phred encoding, of which there are different formats. Most, if not all, new files will use Phred+33 but be careful if you are using older data as it may use a different one. See the "Encoding" section in the following link for more info: https://en.wikipedia.org/wiki/FASTQ_format.
    +  NOTE: __‘@’__ can be used as a quality value.

An example of the information of one sequence/entry is:

__\@Sequence 1__  
__CTGTTAAATACCGACTTGCGTCAGGTGCGTGAACAACTGGGCCGCTTT__  
__+__  
__=<<<=>\@\@\@ACDCBCDAC\@BAA\@BA\@BBCBBDA\@BB\@>CD\@A\@B?B\@\@__

## Basecalling
```{r, fig.align = 'center',out.width= '15%', echo=FALSE }
knitr::include_graphics(path = "figures/guppy.png", auto_pdf = TRUE)
``` 

In short, basecalling is the converting of ONT signals to bases and quality. In other word it is the conversion of Fast5 files to BAM and/or fastq files.

There are many tools to carry out basecalling with ONT data with ONT sequencing machines carrying this out themselves. However, if you are interested in the tools used for this the primary ones are Guppy and Albacore. Unfortunately it is quite hard to find information about these tools unless you own an ONT machine.

Basecalling with Guppy tutorial:

https://denbi-nanopore-training-course.readthedocs.io/en/latest/basecalling/basecalling.html

Basecalling with Albacore tutorial:

https://denbi-nanopore-training-course.readthedocs.io/en/stable/basecalling/basecalling.html

## QC of fastq file
```{r, fig.align = 'center',out.width= '15%', echo=FALSE }
knitr::include_graphics(path = "figures/qc_inspection.png", auto_pdf = TRUE)
``` 

For this tutorial we will carry out quality control and checking on fastq files. Some QC steps can be carried out on summary files and/or BAM files. However we will be using fastq files for a variety of reasons:

- fastq files are the smallest and so these processes will run quicker and the files will take up less storage. Ideal for a tutorial.
- Fastq files are the easiest to work with.
- You will most likely use fastq files in the future for ONT data. 
  - If you get your data sequenced by a genomic centre they will most likely give you your data in demultiplexxed (one fastq file per sample) fastq files.
  - ONT machines generally come with in built basecalling and so will most likely provide fastq files as well as the other formats now.

Before carrying out any specific commands we will first move into the relevant directory.

```{bash eval=FALSE}
cd ~/QC_workshop/ONT_QC/
```

Initialise environment. This needs to be done if you start working in a new terminal. Each terminal will only remember what you have told it.

```{bash eval=FALSE}
. usenanopack-1.1.0
```

Look in the directory called data and you will notice there are a few directories. You can see the contents of all these directories with the below command.

```{bash eval=FALSE}
ls data/*
```

You will notice that each directory has one fastq file. ONT data likes to be organised with data for one sample being in one directory. This is reflected in the software that we will use which likes to be pointed a directory with one sample in it.

To start with we will only use the fastq file within the directory called "Acinetobacter". As you may have figured out this contains ONT sequencing data of an Acinetobacter genome. Specifically the data is a subset of the SRA (Sequence Read Archive) Run: SRR7119550.

### NanoStat
```{r, fig.align = 'center',out.width= '15%', echo=FALSE }
knitr::include_graphics(path = "figures/metrics.png", auto_pdf = TRUE)
``` 

The first step is to acquire stats for the sequences in our fastq file. We will use NanoStat (https://github.com/wdecoster/nanostat).

NanoStat is one of the many tools contained in the NanoPack suite (https://github.com/wdecoster/nanopack). We will also use the tools NanoPlot and NanoFilt.

We want to have a tidy set of directories at the end of this analysis. It would be an untidy mess if we had all the output files in the current directory. We will therefore be making directories and subdirectories to send our output.

```{bash eval=FALSE}
mkdir nanostats
```

Finally we will now run NanoStat. The options used are:

- `-n` : File name/path for the output.
- `-t` : Number of threads the script can use.
- `--fastq` : Input data is in fastq format. Other options that can be used are `--fasta`, `--summary`, and `--bam`.

```{bash eval=FALSE}
NanoStat -n nanostats/Acinetobacter_nanostats.tsv \
-t 4 --fastq ./data/Acinetobacter/Acinetobacter_genome_ONT.fastq
```

Now we can look at the output text file. In this case we will use the convenient `less` command.

```{bash eval=FALSE}
less nanostats/Acinetobacter_nanostats.tsv
```

The file contains four sections with informative headers. These are:

1. __General summary__
   - A list of general summary metrics.
2. __Number, percentage and megabases of reads above quality cutoffs__
   - Based on the mean quality value of a read.
   - The ">Q5" line shows the number and % of reads with a mean quality above Q5. It also shows the total amount of megabases these reads contain.
3. __Top 5 highest mean basecall quality scores and their read lengths__
   - Shows the top 5 reads with the highest mean quality scores.
4. __Top 5 longest reads and their mean basecall quality score__
   - Shows the top 5 longest reads.

### Porechop
```{r, fig.align = 'center',out.width= '60%', echo=FALSE }
knitr::include_graphics(path = "figures/porechop_logo_knife.png", auto_pdf = TRUE)
``` 

Porechop is a tool to find and remove adapters from ONT data (https://github.com/rrwick/Porechop). Adapters are artificial sequences essnetial for sequencing but of no biological value and so you will always want them removed.

Porechop is no longer supported but it still works and there is no alternative for adapter removal. Depending on which basecaller was used on your data, adapter removal may have already been carried out. However, it is always best to run porechop if you are not sure. 

Porechop has a list of known adapters it will look for and remove. These contain:

- Ligation kit adapters
- Rapid kit adapters
- PCR kit adapters
- Barcodes
- Native barcoding
- Rapid barcoding

Porechop will look for these adapters at the start and end of each read. Then it will look for adapters within the sequence (known as middle adapters to Porechop). If it finds a middle adapter it will conclude the read is chimeric (a recombinant read containing sequence from 2 or more reads) and split the read. Depending on the number of middle adpaters the chimeric read may split into 2 or more reads.

With all that explanation we will now run Porechop. Thankfully the command is relatively straight forward with the options:

- `-t`: Number of threads to be used.
- `-i`: Input path of a fasta file, fastq file, or a directory. If a directory is specified it will be recursively searched for fastq files.
- `-o`: Output path. This will either be a fastq or fasta file name.

```{bash eval=FALSE}
#Create output directory
mkdir porechop
#Run porechop command
porechop -t 4 -i ./data/Acinetobacter/Acinetobacter_genome_ONT.fastq \
-o porechop/Acinetobacter.porechop.fastq
```

Porechop will take a while to run. Whilst it is running look at the screen output to get an idea of what it is doing. Once it is finished look at the bottom of the printed results.

Questions:

- How many reads had adapters trimmed from their start?
- How many bases were removed by adapters being trimmed from the end of reads?
- How many read were split based on middle adpaters?

For more uses of Porechop please see the below links:

- https://github.com/rrwick/Porechop#quick-usage-examples
- https://github.com/rrwick/Porechop#full-usage

### NanoPlot
```{r, fig.align = 'center',out.width= '15%', echo=FALSE }
knitr::include_graphics(path = "figures/nanoplot_hexes.png", auto_pdf = TRUE)
``` 

NanoPlot can be thought of as the FastQC for ONT data. It produces a lot of useful visualisations to investigate the quality of ONT sequencing data. It can be used for fastq, fasta, bam, and sequencing summary files. The link for its github page is: https://github.com/wdecoster/NanoPlot

ONT data has much lower quality scores than Illumina with Q10 being good. If you have enough coverage and length the low quality can be corrected by downstream processes not covered in this tutorial.

Prior to running NanoPlot we will make a directory for the NanoPlot output. NanoPlot creates a lot of files so we'll make a subdirectory for the NanoPlot output of porechopped fastq files.

```{bash eval=FALSE}
mkdir nanoplot
mkdir nanoplot/porechop
```

Now to run NanoPlot. The options we will use are:

- `-t`: Number of threads to be used.
- `--fastq`: Specifies the input path which is a fastq file.
- `-o`: Directory where the output will be created.
- `-p`: Prefix of output files. It is useful to have "_" at the end of the prefix.
- `--plots`: Specifies what type of bivariate plot is to be created (more on this later). I find "hex" to be the best. 

```{bash eval=FALSE}
NanoPlot -t 4 \
--fastq porechop/Acinetobacter.porechop.fastq \
-o nanoplot/porechop -p Acinetobacter_ \
--plots hex 
```

You may get below warning. This is fine and can be ignored.

```{bash, eval = FALSE}
/pub37/matt/programs_chos_8/anaconda3/lib/python3.7/_collections_abc.py:702: MatplotlibDeprecationWarning: The global colormaps dictionary is no longer considered public API.
  return len(self._mapping)
/pub37/matt/programs_chos_8/anaconda3/lib/python3.7/_collections_abc.py:720: MatplotlibDeprecationWarning: The global colormaps dictionary is no longer considered public API.
  yield from self._mapping
```

List the files in the output directory.

```{bash, eval=FALSE}
ls nanoplot/porechop/
```

You will notice there are quite a few files. These should all start with "Acinetobacter_" thanks to the `-p` option. To quickly check all the results we can open the produced html file with firefox.

```{bash eval=FALSE}
firefox nanoplot/porechop/Acinetobacter_NanoPlot-report.html
```

The first section contains NanoStat output. Quickly look over this and see how it compares to the NanoStat output of the pre-porechopped reads.

Tip: To open a new terminal, right click VNC background -> Applications -> Shells -> Bash.

After the __Summary Statistics__ section there is a __Plots__ section. This contains the plots:

- __Histogram of read lengths__
   - Histogram of "Number of reads" (y) against "Read length" (x).
- __Histogram of read lengths after log transformations__
   - Histogram of "Number of reads" (y) against log transformed "Read length" (x).
- __Weighted Histogram of read lengths__
   - Histogram of "Number of bases" (y) against "Read length" (x).
- __Weighted Histogram of read lengths after log transformation__
   - Histogram of "Number of bases" (y) against log transformed "Read length" (x).
- __Dynamic histogram of Read length__
   - An interactive Histogram of "Number of reads" (y) against "Read length" (x) produced by plotly (https://plotly.com/).
   - You can zoom into areas by clicking and dragging. Makes boxes to zoom into a specific area. Click and drag left or right only to zoom into a specific part of the x axis. Click and drag up or down only to zoom into a specific part of the y axis. Click the home icon on the top right to reset the axes.
- __Yield by length__
   - Plot showing the "Cumulative yield for minimal length" (y) by "Read length" (x).
   - The cumulative yield is measured in Gigabases (billion bases).
   - This plot is useful to know how many bases you would retain if you filtered reads based on read length.
- __Read lengths vs Average read quality plot using hexagonal bins__
   - This is the most informative plot. It is the bivariate plot we indicated we wanted as hexagonal with the option `--plots hex`.
   - Each hex is a bin with a darker colour representing more reads.
   - At the top of the plot is a histogram of Number of reads against read lengths.
   - At the right of the plot is a sideways histogram of number of reads against average read quality.
   
Questions to be answered with the html file:

- How long is the longest read?
- What is the highest mean basecall quality score of a read?
- How many reads have a mean basecall quality score >Q10?
- Approximately where is the highest density of reads in terms of read lengths and average read quality.
- Roughly, do the longer reads have relatively high, medium, or low mean qualities?

Using different input files will give you different plots. For more details please see the "Plots Generated" section on: https://github.com/wdecoster/NanoPlot.

### NanoFilt
```{r, fig.align = 'center',out.width= '10%', echo=FALSE }
knitr::include_graphics(path = "figures/quality_filtering.png", auto_pdf = TRUE)
```

NanoFilt can be used to remove/filter reads by quality and/or read length (https://github.com/wdecoster/nanofilt). This is very useful as you will most likely want long and good quality reads for downstream processes, such as genome assemblies.

It is always important to know what you data is and if your planned filtering and triming is appropriate. FOr example you may be working with amplicon data where the read lenghts will vary between 500bp and 750bp. In that case it is useful to set a min length of 500 and a maxlength of 750.

This data is from a genome so wanting long and high quality reads is appropriate. Overall with NanoPlot the quality looks good. The main exception is the amount of sequencing/the coverage but this is because it is a subset of the whole data for workshop purposes. There are some shorter and lower quality reads which we will want removed.

First we will create an output directory.
```{bash eval=FALSE}
mkdir nanofilt
```

We will filter out sequences that are shorter than 500bp (`-l 500bp`) and filter out sequences with a average read quality less than Q10 (`-q 10`. We have chosen these values as they are a good default for genomic data. Q10 appeared to be a good choice from NanoPLot as the vast majorit of reads had a quality of >Q10.

```{bash eval=FALSE}
cat porechop/Acinetobacter.porechop.fastq | \
NanoFilt -l 500 -q 10 > \ nanofilt/Acinetobacter.nanofiltq10minlen500.porechop.fastq
```

Are the chosen options appropriate? Let us find out.

### Final check
```{r, fig.align = 'center',out.width= '15%', echo=FALSE }
knitr::include_graphics(path = "figures/finish.png", auto_pdf = TRUE)
```

After trimming or filtering reads (quality control) it is always important to carry out quality check. We will therefore run NanoPlot again.

Note: We are using a long informative output directory name. This is important as we may need to rerun NanoFilt a few more times with different options until we are happy. 

```{bash eval=FALSE}
#Make an output directory before running NanoPlot
mkdir nanoplot/nanofiltq10minlen500_porechop
#Run NanoPlot on the filtered and porechopped data
NanoPlot -t 4 \
--fastq nanofilt/Acinetobacter.nanofiltq10minlen500.porechop.fastq \
-o nanoplot/nanofiltq10minlen500_porechop/ -p Acinetobacter_ \
--plots hex 
```

