# Porechop {#porechopchap}
<center>
![](figures/porechop_logo_knife.png){style="width:600px; border-radius:15px; border:5px solid white; background:null"}
</center>

`Porechop` is a tool to find and remove adapters from ONT data (https://github.com/rrwick/Porechop). Adapters are artificial sequences essential for sequencing but of no biological value and so you will typically want them removed.

`Porechop` is no longer supported but it still works and is a good tool. Depending on which basecaller was used on your data, adapter removal may have already been carried out. However, it is always best to run `porechop` if you are not sure. 

`Porechop` has a list of known adapters it will look for and remove. These contain:

- Ligation kit adapters
- Rapid kit adapters
- PCR kit adapters
- Barcodes
- Native barcoding
- Rapid barcoding

`Porechop` will look for these adapters at the start and end of each read. Then it will look for adapters within the sequence (known as middle adapters to `Porechop`). If it finds a middle adapter it will conclude that the read is chimeric (a recombinant read containing sequences from 2 or more reads) and split the read. Depending on the number of middle adapters the chimeric read may split into 2 or more reads.

## Porechop: run
<center>
![](figures/enter_key_blue.png){style="width:150px"}
</center>

With all that explanation we will now run `Porechop.` Thankfully the command is relatively straight forward with the options:

- `-t`: Number of threads to be used.
- `-i`: Input path of a fasta file, fastq file, or a directory. If a directory is specified it will be recursively searched for fastq files.
- `-o`: Output path. This will either be a fastq or fasta file name.

```{bash eval=FALSE}
#Create output directory
mkdir porechop
#Run porechop command
porechop -t 4 -i ./data/Acinetobacter/Acinetobacter_genome_ONT.fastq \
-o porechop/Acinetobacter.porechop.fastq
```

`Porechop` will take a while to run. Whilst it is running, look at the screen output to get an idea of what it is doing. When finished, look at the bottom of the printed results.

## Porechop: MCQs
<center>
![](figures/question_bubble_blue.png){style="width:150px"}
</center>

```{r, echo = FALSE}
opts_p <- c("__0__", answer="__1,952__", "__79,015__")
```
1. How many reads had adapters trimmed from their start? `r longmcq(opts_p)`

```{r, echo = FALSE}
opts_p <- c("__0__", "__1,952__", answer="__79,015__")
```
2.  How many bases were removed by adapters being trimmed from the end of reads? `r longmcq(opts_p)`

```{r, echo = FALSE}
opts_p <- c(answer="__0__", "__1,952__", "__79,015__")
```
3. How many reads were split based on middle adapters? `r longmcq(opts_p)`

## Porechop: summary
<center>
![](figures/sum_blue.png){style="width:150px"}
</center>

`Porechop` did its job and removed adapters successfully. The next step is to view the quality of our reads so we can be informed when we filter out poor quality reads and bases.

For more uses of `Porechop` please see the below links:

- https://github.com/rrwick/Porechop#quick-usage-examples
- https://github.com/rrwick/Porechop#full-usage