# NanoPlot {#nanoplotchap}
```{r, fig.align = 'center',out.width= '15%', echo=FALSE }
knitr::include_graphics(path = "figures/nanoplot_hexes.png", auto_pdf = TRUE)
``` 

`NanoPlot` can be thought of as the `fastqc` for ONT data. It produces a lot of useful visualisations to investigate the quality of ONT sequencing data. It can be used for __fastq__, __fasta__, __BAM__, and sequencing summary files. The link for its github page is: https://github.com/wdecoster/NanoPlot

ONT data has much lower quality scores than Illumina with Q10 being good. If you have enough coverage and length the low quality can be corrected by downstream processes not covered in this tutorial.

## NanoPlot: run
```{r, fig.align = 'center',out.width= '15%', echo=FALSE }
knitr::include_graphics(path = "figures/enter_key_green.png", auto_pdf = TRUE)
```

Prior to running `NanoPlot` we will make a directory for the `NanoPlot` output. As `NanoPlot` creates a lot of files, we'll make a subdirectory for the `NanoPlot` output for the porechopped data.

```{bash eval=FALSE}
mkdir nanoplot
mkdir nanoplot/porechop
```

Now to run `NanoPlot`. The options we will use are:

- `-t`: Number of threads to be used.
- `--fastq`: Specifies the input path which is a fastq file.
- `-o`: Directory where the output will be created.
- `-p`: Prefix of output files. It is useful to have "_" at the end of the prefix.

```{bash eval=FALSE}
NanoPlot -t 4 \
--fastq porechop/Acinetobacter.porechop.fastq \
-o nanoplot/porechop -p Acinetobacter_ 
```

## NanoPlot: output
```{r, fig.align = 'center',out.width= '15%', echo=FALSE }
knitr::include_graphics(path = "figures/output_file_green.png", auto_pdf = TRUE)
``` 

List the files in the output directory.

```{bash, eval=FALSE}
ls nanoplot/porechop/
```

There are quite a few files. These should all start with \"Acinetobacter_\" thanks to the `-p` option. To quickly check all the results we can open the report html file with firefox.

```{bash eval=FALSE}
firefox nanoplot/porechop/Acinetobacter_NanoPlot-report.html
```

The first section contains [`NanoStat` output](#nanostat_output). Quickly look over this and see how it compares to the `NanoStat` output of the pre-porechopped reads.

`r hide("NanoStat of pre-porechopped reads")`
```{r, echo=FALSE}
#Create dataframe of the nanostat pre-porechopped data
df <- read.csv("files/Acinetobacter_nanostats.tsv", 
               sep = ":",
               check.names=FALSE)
df[,2] <- trimws(df[,2],which="both", whitespace="[ \t\r\n]")
#
colnames(df) <- c("General summary","values")
#Display table with kable
library("knitr")
kable(df)
#Display table with flextable
#https://ardata-fr.github.io/flextable-book/index.html
# library("flextable")
# ft <- flextable(df)
# ft <- autofit(ft)
# ft
```
`r unhide()`

After the __Summary Statistics__ section there is a __Plots__ section. 
These are all plots created by [plotly](https://plotly.com/) meaning they are interactive.
With these plots you can:

  - Make boxes to zoom into a specific area. 
  - Click and drag left or right only to zoom into a specific part of the x axis. 
  - Click and drag up or down only to zoom into a specific part of the y axis. 
  - Click the home icon (`r icons::ionicons("home")`) on the top right to reset the axes.
  - Hover over a points/bars to see the specific values of it.

The plots are:

- __Weighted Histogram of read lengths__
   - Histogram of \"Number of _bases_\" (y) against \"Read length\" (x).
- __Weighted Histogram of read lengths after log transformation__
   - Histogram of \"Number of _bases_\" (y) against log transformed \"Read length\" (x).
- __Non weighted histogram of read lengths__
   - Histogram of \"Number of _reads_\" (y) against \"Read length\" (x).
- __Non weighted histogram of read lengths after log transformations__
   - Histogram of \"Number of _reads_\" (y) against log transformed \"Read length\" (x).
- __Yield by length__
   - Plot showing the \"Cumulative yield for minimal length\" (y) by \"Read length\" (x).
   - The cumulative yield is measured in Gigabases (billion bases).
   - This plot is useful to know how many bases you would retain if you filtered reads based on read length.
- __Read lengths vs Average read quality plot using dots__
   - Each dot represents a single read.
   - At the top of the plot is a histogram of Number of reads against read lengths.
   - At the right of the plot is a sideways histogram of number of reads against average read quality.
- __Read lengths vs Average read quality kde plot__
   - Each dot represents a single read with a topology map showing the density of plots. THe darker the area the more dense.
   - This is the most informative plot.
   - At the top of the plot is a histogram of Number of reads against read lengths.
   - At the right of the plot is a sideways histogram of number of reads against average read quality.

## NanoPlot: MCQs
```{r, fig.align = 'center',out.width= '15%', echo=FALSE }
knitr::include_graphics(path = "figures/question_bubble_green.png", auto_pdf = TRUE)
```
Please attempt the below MCQs with the information in the report html.

```{r, echo = FALSE}
opts_p <- c(answer="__9.5__", "__6,962.1__", "__8.8Mb__")
```
1. What is the mean read quality for all the data?`r longmcq(opts_p)`

```{r, echo = FALSE}
opts_p <- c("__9.5__", answer="__6,962.1__", "__8.8Mb__")
```
2.  What is the mean read length of the data? `r longmcq(opts_p)`

```{r, echo = FALSE}
opts_p <- c("__9.5__", "__6,962.1__", answer="__8.8Mb__")
```
3.  How many Megabases are within reads with a quality cutoff of >Q10?  `r longmcq(opts_p)`

```{r, echo = FALSE}
opts_p <- c("__10__", "__146__", answer="__4Mb (0.004Gb)__")
```
4. Approximately how many bases would you retain if you filtered out/removed reads with a length <10,000bp (Yield by length)? `r longmcq(opts_p)`

```{r, echo = FALSE}
opts_p <- c("__6-8__", "__8-10__", answer="__10-12__")
```
5. Look at the "Read lengths vs Average read quality plot". In what average quality range are the majority of reads? `r longmcq(opts_p)`

## NanoPlot: summary
```{r, fig.align = 'center',out.width= '15%', echo=FALSE }
knitr::include_graphics(path = "figures/sum_green.png", auto_pdf = TRUE)
```

We have use `NanoPlot` to produce read quality and length stats in table and plot form. Our data still looks good after porechopping but we still need to carry out some filtering.

You can use `NanoPlot` with different input files. This will give you different plots. For more details please see the \"Plots Generated\" section on: https://github.com/wdecoster/NanoPlot.
