# PacBio filtering {#pbfiltchap}
<center>
![](figures/quality_filtering.png){style="width:150px; background: white; border-radius:5px; border: white solid 5px"}
</center>

PacBio HiFi reads are generated via CCS analysis which computes a consensus sequence from multiple passes of the circularised SMRTbell molecules. Each 'pass' will be interspersed with an adapter sequence (as shown in 10.1). Whilst the majority of these adapter sequences are removed by the CCS analysis that generates the HiFi BAM file, it is good practice to check for any remaining sequences to ensure they are removed before downstream analysis.

We will perform this using the software [HiFiAdapterFilt](https://bmcgenomics.biomedcentral.com/articles/10.1186/s12864-022-08375-1). 

This software has already been installed, but we will start by setting the PATH variable.

## $PATH

The PATH environment variable is a list of directories that an operating system searches to find executable files and commands (programs). If a directory is included in the `$PATH`, it won’t be necessary to type out the full name of the executable when running a command. Directories can be added to the path on a temporary basis using the export command, but if you close your terminal or start a new session you will have to rerun the export command.

```{bash eval=FALSE}
# Add the hifiadapterfilt script and the database to your path

export PATH=$PATH:/pub14/tea/nsc206/git_installs/HiFiAdapterFilt

export PATH=$PATH:/pub14/tea/nsc206/git_installs/HiFiAdapterFilt/DB
```

The echo command displays text on the screen, and we can use it to show the contents of the `$PATH` environment variable now that we have exported the locations of hifiadapterfilt and hifiadapterfilt/db.

```{bash eval=FALSE}
# Check the PATH environment variable with echo

echo $PATH
```

Great, now we can proceed to adapter filtering.

## HiFiAdapterFilt

<center>
![](figures/enter_key_purple.png){style="width:150px; ; background: white; border-radius:5px; border: white solid 5px"}
</center>

[HiFiAdapterFilt](https://github.com/sheinasim/HiFiAdapterFilt) blasts the reads against a database of PacBio adapter sequences and excludes reads if the match with an adapter is good enough (the default is a blast percent identity match of 97% and a minimum length of 44 bp). It creates a file with the suffix ‘.filt.fastq.gz’ with the adapters removed that can be used for QC and downstream analysis.

Run HiFiAdapterFilt on the fastq file:

```{bash eval=FALSE}
hifiadapterfilt.sh -p ERR13246467_qc
```

The `-p` parameter sets the prefix of the file to be filtered.

Use `less` to inspect the results to check for adapter contamination:

```{bash eval=FALSE}
less ERR13246467_qc.stats
```

```{r, echo = FALSE}
opts_p <- c("__99.99%__", "__2__", answer="__13798__")
```
How many CCS reads are in the unfiltered dataset?  `r longmcq(opts_p)`

```{r, echo = FALSE}
opts_p <- c("__99.99%__", answer="__2__", "__13798__")
```
How many adapter contaminated reads were found?  `r longmcq(opts_p)`

```{r, echo = FALSE}
opts_p <- c(answer="__99.99%__", "__2__", "__13798__")
```
What percentage of CCS reads are retained (i.e. unfiltered)?  `r longmcq(opts_p)`

Very few reads were found to contain adapter sequences. This isn't entirely unexpected as the CCS algorithm on the instrument will filter out most of them. However, the authors of HiFiAdapterFilt found that >95% of published HiFi datasets did contain some adapter sequences, so it is worthwile to check and remove these.

## PacBio adapter filtering: summary
<center>
![](figures/sum_purple.png){style="width:150px; background: white; border-radius:5px; border: white solid 5px"}
</center>

We have now filtered any remaning sequencing adapters from our HiFi data. Next we will learn how to QC the data.
