# CLR: Filtering
```{r, fig.align = 'center',out.width= '10%', echo=FALSE }
knitr::include_graphics(path = "figures/quality_filtering.png", auto_pdf = TRUE)
``` 

Our quality check of the PacBio data doesn\'t include any assessment of sequence quality. Unaligned PacBio data doesn\'t have a quality score in the same way as the Illumina data we looked at on Tuesday. Quality scores for PacBio are generated when the reads are aligned - either to a reference sequence or subreads aligned to each other to generate CCS reads. However, it is often a good idea to use just the longest PacBio reads for a _de novo_ assembly.

To filter the data by minimum CLR length we will use SequelTools option `-t F`.

Other options used are:

- `-u` to specify the file containing the list of subread .bam files
- `-c` to specify the file containing the list of scraps .bam files
- `-C` to specify to filter by minimum CLR length
- `-Z` to specify the minimum length to keep a CLR
- `-f` to specify the format of the output files: `s`=sam, `b`=bam, `2`=both
- `-o` to specify the name of an output folder for the filtered data files

First we need to make a fofn for the scraps.bam files, as these are needed when using the filtering tool:
```{bash eval=FALSE}
find $(pwd) -name "*scraps.bam" | sort > CLR_scraps.txt
```

Then we can run the filtering command. Here we will keep only CLRs of at least 10,000 bp:
```{bash eval=FALSE}
SequelTools.sh -t F -u CLR_subreads.txt -c CLR_scraps.txt \
-C -Z 10000 -f b -o filtered
```

When this has finished running the \'filtered\' folder will contain filtered BAM files for all three samples. Move into this folder using `cd`:
```{bash eval=FALSE}
cd filtered
```

From here let\'s run the quality control tool on the filtered files and compare the plots to those from the raw files. First we need a new fofn for the filtered subread files:
```{bash eval=FALSE}
find $(pwd) -name "*subreads.bam" | sort > filt_subreads.txt
```

Then run the quality control tool again:
```{bash eval=FALSE}
SequelTools.sh -t Q -p a -u filt_subreads.txt -o filt_QC
```

When this has finished running the \'filt_QC\' folder will contain a text file of summary values and a series of plots saved as pdf files. We can view these as before, for example:
```{bash eval=FALSE}
firefox filt_QC/subreadSizesBoxplots.pdf
```

And compare this to the boxplots from the unfiltered data:
```{bash eval=FALSE}
firefox ~/QC_workshop/PB_QC/data/CLR/CLR_QC/subreadSizesBoxplots.pdf
```

The difference looks quite subtle on the plots as the size range of reads is large, but the subread N50 value (indicated by the blue diamond) has increased by ~2,000 bp in the filtered data for Sample2 and Sample3.

There are also still subreads present in the data that are < 10,000 bp. This is because we have filtered out CLRs that are < 10,000 bp, but a CLR > 10,000bp can still be made up of subreads smaller than this.

This is definitely the case for Sample1 and Sample2, where the DNA fragment length of the library was shorter than the CLR lengths, hence there are subreads < 10,000 bp within longer CLRs.

Sample3 shows a more obvious difference before and after filtering:
```{bash eval=FALSE}
firefox ~/QC_workshop/PB_QC/data/CLR/CLR_QC/Sample3.readLenHists.pdf \
filt_QC/Sample3_flt.readLenHists.pdf
```

For this sample the DNA fragment length of the library was much greater than 10,000 bp, and so the majority of CLRs consist of just one pass of this long insert. The large number of short subreads remaining in the filtered file (blue bars) represent subreads only partially sequenced on the \'return journey\' of the circular library DNA molecule. These are not present when we look only at the longest subread present in each CLR (green bars).

SequelTools doesn\'t have an option for filtering subreads by length. To do this we can convert the filtered BAM file to a fastq or fasta file and use another downstream tool. The fastq file will not contain any useful information on sequence quality (every base is assigned a quality character of \'!\' which is equal to zero), but this file format is useful if only fastq format is accepted as input for a downstream tool.

To do this use the commands `pbindex` and `bam2fastq` from the smrttools package:
```{bash eval=FALSE}
pbindex Sample1_flt.subreads.bam
```

Note that you may see a long warning message ending with \"No such file or directory\". However, if you run `ls` you should see that the output file has been generated.

This command generates a .pbi index file that is needed before we can convert the BAM file to fastq format. The `bam2fastq` command requires an output file prefix name, specified by `-o`, and the name of the BAM file we want to convert:
```{bash eval=FALSE}
bam2fastq -o Sample1_filtered Sample1_flt.subreads.bam
```

You should now have a compressed fastq file for the CLR-filtered subreads from Sample1.