# (APPENDIX) Appendix {-}

# Further resources
<center>
![](figures/further_resources.png){style="width:200px; background: white; border-radius:50px"}
</center>

- [Long read tools](https://long-read-tools.org/index.html)

## ONT

Unfortunately with ONT you need an account to access a lot of the content on its website.

- [Nanopack overview](https://github.com/wdecoster/nanopack)
- [ONT website](https://nanoporetech.com/)

## PacBio

- [PacBio website](https://www.pacb.com/)
- [PacBio HiFi sequencing](https://www.pacb.com/technology/hifi-sequencing/)
- [PacBio HiFI reads brochure](https://www.pacb.com/wp-content/uploads/Application-Brochure-HiFi-Reads-for-Highly-Accurate-Long-read-Sequencing.pdf)
- [Sequel II and IIe data files](https://www.pacb.com/wp-content/uploads/Sequel_II_and_IIe_Data_Files.pdf)
- [Revio brochure](https://www.pacb.com/wp-content/uploads/Revio-brochure.pdf)

# Mamba installs {#mamba_install}
<center>
![](figures/mamba_logo.png){style="width:200px; border-radius:15px; background:null"}
</center>

## Mamba installation {#mamba_install}

Mamba is a reimplementation of conda. It is a great tool for installing bioinformatic packages including R packages.

Mamba github: <https://github.com/mamba-org/mamba>

The best way to use Mamba is to install Miniforge. It has both Conda and Mamba commands.

Miniforge installation: <https://github.com/conda-forge/miniforge>

Mamba guide: <https://mamba.readthedocs.io/en/latest/user_guide/mamba.html>

## ONT QC environment
To create the mamba environment `ont_qc` run the below commands in your bash. You will need to have `mamba` installed first.

```{bash, eval=FALSE}
#Oxford Nanopore QC
mamba create -n ont_qc
mamba activate ont_qc
#Install tools
mamba install -c bioconda porechop nanoplot chopper nanostat
```

## PB QC environment
To create the mamba environment `pb_qc` run the below commands in your bash. You will need to have `mamba` installed first.

```{bash, eval=FALSE}
#PacBio QC
mamba create -n pb_qc
mamba activate pb_qc
#Install smrttools
mamba install -c hcc smrtlink-tools
#Install git & make for below installs
mamba install -c anaconda git
mamba install -c anaconda make
#R needed for Sequeltools
mamba install -c conda-forge r-base
```

Install [HiFiAdapterFilt](https://github.com/sheinasim/HiFiAdapterFilt) and [RabbitQCPlus](https://github.com/RabbitBio/RabbitQCPlus) via `git`.
I suggest creating a directory called "git_installs" in your home directory and running this code there.

```{bash, eval=FALSE}
#git_installs directory
mkdir ~/git_installs
cd ~/git_installs
#HiFiAdapterFilt
git clone https://github.com/sheinasim/HiFiAdapterFilt
#RabbitQCPlus
git clone https://github.com/RabbitBio/RabbitQCPlus.git
cd RabbitQCPlus
make -j4
```

To run these programs in your own machines you will need to use the full path of the command files.
If you installed these in your "~/git_installs" examples are below.

```{bash, eval=FALSE}
#hifiadapterfilt.sh help manual
~/git_installs/HiFiAdapterFilt/hifiadapterfilt.sh --help
#RabbitQCPlus help manual
~/git_installs/RabbitQCPlus/RabbitQCPlus --help
```
