--- 
title: "Long read Quality Control"
author: "Matthew Gemmell & Helen Hipperson"
date: "`r Sys.Date()`"
site: bookdown::bookdown_site
output: bookdown::gitbook
documentclass: book
bibliography: [book.bib, packages.bib]
biblio-style: apalike
link-citations: yes
---

# Introduction
```{r, fig.align = 'center',out.width= '20%', echo=FALSE }
knitr::include_graphics(path = "figures/well.png", auto_pdf = TRUE)
``` 

This practical session aims to introduce you to Long read Quality Control. The topics covered are:

- Acquiring the workshop data
- Example of PacBio quality control
- Example of ONT quality control

<!--chapter:end:01-Long_read_QC.Rmd-->

# Introduction
```{r, fig.align = 'center',out.width= '20%', echo=FALSE }
knitr::include_graphics(path = "figures/Play.png", auto_pdf = TRUE)
``` 

## The data

The data for today is in the "QC_workshop" directory that you copied into your home directory ("~") in the Illumina QC day.

If you do not have this directory the command to copy it to your home directory is below.

```{bash eval=FALSE}
cp -r /pub39/tea/matthew/NEOF/QC_workshop/ ~
```

<!--chapter:end:02-Intro.Rmd-->

# PacBio

```{r, fig.align = 'center',out.width= '40%', echo=FALSE }
knitr::include_graphics(path = "figures/Pacbio_logo.png", auto_pdf = TRUE)
``` 

<!--chapter:end:03-PB.Rmd-->

# ONT
```{r, fig.align = 'center',out.width= '60%', echo=FALSE }
knitr::include_graphics(path = "figures/ONT_logo.png", auto_pdf = TRUE)
``` 

In this section we are going to carry out a quick QC of ONT data using the tool suite NanoPack and the tool Porechop. We will use a the fastq files 

## File formats
```{r, fig.align = 'center',out.width= '15%', echo=FALSE }
knitr::include_graphics(path = "figures/files.png", auto_pdf = TRUE)
``` 

### Fast5
```{r, fig.align = 'center',out.width= '15%', echo=FALSE }
knitr::include_graphics(path = "figures/fast5.png", auto_pdf = TRUE)
``` 

The raw data from ONT machines come as Fast5 (.fast5) files. This contains the signal data from the pore which can be processed into more useful files.

Fast5 files can contain:

+ Raw signal data
+ Run metadata
+ fastq-basecalls
+ Other additional analyses

The Fast5 file format is an implementation of the HDF5 file format specifically for ONT sequencing data. For more information on Fast5 and a tool to interface with them see:
https://github.com/nanoporetech/ont_fast5_api

### Summary file

The MinION and GridION output a single sequencing summary file in addition to the Fast5 files. This file contains metdata which descirbes each sequenced read.

We will not use Fast5 or summary files for this tutorial as they are not needed most of the time. It is likely these files will not be intially provided to you by a sequencing service as they are large and not often needed. However, if you do require them you can always ask but be careful with how long the sequencing centre will retain your data.

### BAM file
```{r, fig.align = 'center',out.width= '15%', echo=FALSE }
knitr::include_graphics(path = "figures/bam.png", auto_pdf = TRUE)
``` 

BAM files (.bam) are the binary version of SAM files (.sam). This means:

- BAM files are not human readable whilst SAM files are.
- SAM files are larger than BAM.

Generally programs can work on both BAM and SAM files and it is preferable to store data in BAM format over SAM. Even though BAM files are smaller than their matching SAM file they are still very large and can be > 100GB.

SAM stands for "Sequence Alignment/Map" (the B in BAM is Binary). SAM files are tab delimited and contain alignment information. 

It can be useful to contain unaligned reads from sequencing machines (e.g. PacBio and ONT) as they can contain more metadata in the header and per-record auxiliary tags. 

If you are working with SMA/BAM files the following link will prove useful: 
https://www.htslib.org/

For more information on the SAM and BAM format please see: 
https://samtools.github.io/hts-specs/SAMv1.pdf


### fastq file
```{r, fig.align = 'center',out.width= '15%', echo=FALSE }
knitr::include_graphics(path = "figures/fastq.png", auto_pdf = TRUE)
``` 

The fastq file format is very consistent and so there is no real difference between fastq files for Illumina, PacBio, and ONT data.

All fastq files will contain a number of lines divisible by 4. This is because each entry/sequence will take up four lines consisting of the following information:

1. Header for fastq entry known as the fastq header. 
    + This always begins with a __‘@’__
    + This is where you might see the most difference between different data. Different machines and different public databases will use different formats for fastq headers.
2. Sequence content
3. Quality header 
    + Always begins with a ‘+’. Sometimes also contains the same information as fastq header.  
4. Quality
    + Each base in the 2nd line will have a corresponding quality value in this line.
    + Note this uses Phred encoding, of which there are different formats. Most, if not all, new files will use Phred+33 but be careful if you are using older data as it may use a different one. See the "Encoding" section in the following link for more info: https://en.wikipedia.org/wiki/FASTQ_format.
    +  NOTE: ‘@’ can be used as quality values.

An example of the information of one sequence/entry is:

__\@Sequence 1__  
__CTGTTAAATACCGACTTGCGTCAGGTGCGTGAACAACTGGGCCGCTTT__  
__+__  
__=<<<=>\@\@\@ACDCBCDAC\@BAA\@BA\@BBCBBDA\@BB\@>CD\@A\@B?B\@\@__

## Basecalling
```{r, fig.align = 'center',out.width= '15%', echo=FALSE }
knitr::include_graphics(path = "figures/guppy.png", auto_pdf = TRUE)
``` 

In short, basecalling is the converting of ONT signals to bases and quality. In other word it is the conversion of Fast5 files to BAM and/or fastq files.

There are many tools to carry out basecalling with ONT data with ONT sequencing machines carrying this out themselves. However, if you are interested in the tools used for this the primary ones are Guppy and Albacore with others. Unfortunately it is quite hard to find information about these tools unless you own an ONT machine.

Basecalling with Guppy tutorial:

https://denbi-nanopore-training-course.readthedocs.io/en/latest/basecalling/basecalling.html

Basecalling with Albacore tutorial:
https://denbi-nanopore-training-course.readthedocs.io/en/stable/basecalling/basecalling.html

## QC of fastq file
```{r, fig.align = 'center',out.width= '15%', echo=FALSE }
knitr::include_graphics(path = "figures/qc_inspection.png", auto_pdf = TRUE)
``` 

For this tutorial we will carry out quality control and checking on fastq files. Some QC steps can be carried out on summary files and/or BAM files. However we will be using fastq files for a variety of reasons:

- fastq files are the smallest and so these processes will run quicker and the files will take up less storage. Ideal for a tutorial.
- You are most likely to be more familiar with fastq files and they are the easiest to work with.
- You will most likely use fastq files in the future for ONT data. 
  - If you get your data sequenced by a genomic centre they will most likely give you your data in demultiplexxed fastq files.
  - ONT machines generally come with in built basecalling and so will most likely provide fastq files as well as the other formats now.

Before carrying out any specific commands we will first move into the relevant directory.

```{bash eval=FALSE}
cd ~/QC_workshop/ONT_QC/
```

Initialise environment. This needs to be done if you start working in a new terminal. Each terminal will only remember what you have told it.

```{bash eval=FALSE}
. usenanopack-1.1.0
```

Look in the directory called data and you will notice there are a few directories. You can see the contents of all these directories with the below command.

```{bash eval=FALSE}
ls data/*
```

You will notice that each directory has one fastq file. ONT data likes to be organised with data for one sample being in one file. This is reflected in the ONT software we will use that likes you to point it to a directory with one sample in it.

To start with we will only use the fastq file within the directory called "Acinetobacter". As you may have figured out this contains ONT sequencing data of an Acinetobacter genome. Specifically the data is a subset of the SRA (Sequence Read Archive) Run: SRR7119550.

### NanoStat
```{r, fig.align = 'center',out.width= '15%', echo=FALSE }
knitr::include_graphics(path = "figures/metrics.png", auto_pdf = TRUE)
``` 

First step is to acquire stats for the sequences in our fastq file. We will use NanoStat (https://github.com/wdecoster/nanostat).

NanoStat is one of the scripts of NanoPack (https://github.com/wdecoster/nanopack). We will also use the scripts NanoPlot and NanoFilt.

We want to have a tidy set of directories at the end of this analysis. It would be a large mess to have all the output files in the current directory. We will therefore be making directories and subdirectories to send our output.

```{bash eval=FALSE}
mkdir nanostats
```

Finally we will now run NanoStat. The options used are:

- `-n` : File name/path for the output.
- `-t` : Number of threads the script can use.
- `--fastq` : zData is in fastq format. Other options that can be used instead are `--fasta`, `--summary`, and `--bam`.

```{bash eval=FALSE}
NanoStat -n nanostats/Acinetobacter_nanostats.tsv \
-t 4 --fastq ./data/Acinetobacter_genome_ONT.fastq
```

Now we can look at the output with any software in Linux that can view text files. In this case we will use the convenient `less`.

```{bash eval=FALSE}
less nanostats/Acinetobacter_nanostats.tsv
```

The files contains four section with informative headers. These are:

1. General summary
  - A list of general summary metrics.
2. Number, percentage and megabases of reads above quality cutoffs
  - Based on the mean quality value of a read.
3. Top 5 highest mean basecall quality scores and their read lengths
  - Shows the top 5 reads based on mean quality score.
4. Top 5 longest reads and their mean basecall quality score
  - Shows the top 5 reads based on read length.

### Porechop
```{r, fig.align = 'center',out.width= '60%', echo=FALSE }
knitr::include_graphics(path = "figures/porechop_logo_knife.png", auto_pdf = TRUE)
``` 

https://github.com/rrwick/Porechop

(Intro to porechop and useage
Mention here that guppy(?) carries out porechop
Porechop is no longer supported but additionally there is no alternative)

```{bash eval=FALSE}
mkdir porechop
porechop -t 4 -i data/Acinetobacter -o porechop/Acinetobacter.porechop.fastq
```

### NanoPlot
```{r, fig.align = 'center',out.width= '15%', echo=FALSE }
knitr::include_graphics(path = "figures/nanoplot_hexes.png", auto_pdf = TRUE)
``` 

https://github.com/wdecoster/NanoPlot

Prior to running NanoPlot we will make a directory for NanoPlot output. NanoPlot makes a lot of files so within that directory we'll make a subdirectory for the NanoPlot output of porechopped fastq files.

```{bash eval=FALSE}
mkdir nanoplot
mkdir nanoplot/porechop
```

Now to run NanoPlot.

```{bash eval=FALSE}
NanoPlot -t 4 \
--fastq porechop/Acinetobacter.porechop.fastq \
-o nanoplot/porechop -p Acinetobacter_ \
--plots hex 
```

May get below error which is fine:

```{bash, eval = FALSE}
/pub37/matt/programs_chos_8/anaconda3/lib/python3.7/_collections_abc.py:702: MatplotlibDeprecationWarning: The global colormaps dictionary is no longer considered public API.
  return len(self._mapping)
/pub37/matt/programs_chos_8/anaconda3/lib/python3.7/_collections_abc.py:720: MatplotlibDeprecationWarning: The global colormaps dictionary is no longer considered public API.
  yield from self._mapping
```

Check HMTL plot

```{bash eval=FALSE}
firefox nanoplot/porechop/Acinetobacter_NanoPlot-report.html
```

### NanoFilt
```{r, fig.align = 'center',out.width= '10%', echo=FALSE }
knitr::include_graphics(path = "figures/quality_filtering.png", auto_pdf = TRUE)
```

```{bash eval=FALSE}
mkdir nanofilt
```

```{bash eval=FALSE}
cat porechop/Acinetobacter.porechop.fastq | \
NanoFilt -l 500 -q 10 > \ nanofilt/Acinetobacter.nanofiltq10minlen500.porechop.fastq
```


### Final check
```{r, fig.align = 'center',out.width= '15%', echo=FALSE }
knitr::include_graphics(path = "figures/finish.png", auto_pdf = TRUE)
```

```{bash eval=FALSE}
NanoStat -t 4 \
-n nanostats/Acinetobacter_nanofiltq10minlen500_porechop_nanostats.tsv \
--fastq nanofilt/Acinetobacter.nanofiltq10minlen500.porechop.fastq
```

```{bash eval=FALSE}
mkdir nanoplot/nanofiltq10minlen500_porechop

NanoPlot -t 4 \
--fastq nanofilt/Acinetobacter.nanofiltq10minlen500.porechop.fastq \
-o nanoplot/nanofiltq10minlen500_porechop/ -p Acinetobacter_ \
--plots hex 
```


<!--chapter:end:04-ONT.Rmd-->

