--- 
title: "Long read Quality Control"
author: "Matthew R. Gemmell & Helen Hipperson"
date: "`r Sys.Date()`"
site: bookdown::bookdown_site
output: bookdown::gitbook
documentclass: book
bibliography: [book.bib, packages.bib]
biblio-style: apalike
link-citations: yes
favicon: figures/NEOF_favicon.png
description: NEOF book for the Long read quality control workshop
cover-image: "figures/NEOF.png"
---
```{r include=FALSE, cache=FALSE}
suppressPackageStartupMessages({
  library(webexercises)
})

knitr::knit_hooks$set(webex.hide = function(before, options, envir) {
  if (before) {
    if (is.character(options$webex.hide)) {
      hide(options$webex.hide)
    } else {
      hide()
    }
  } else {
    unhide()
  }
})
```

```{r cite-packages, include = FALSE}
# automatically create a bib database for R packages
# add any packages you want to cite here
knitr::write_bib(c(
  .packages(), 'bookdown', 'webexercises'
), 'packages.bib')
```

```{r, echo=FALSE}
#This book uses ionicons icons using the icons package
#https://github.com/mitchelloharawild/icons
#To set this up on your rstudio use the following commands
# install.packages("remotes")
#remotes::install_github("mitchelloharawild/icons")
#download_ionicons()
```

```{r, fig.align = 'center',out.width= '30%', echo=FALSE }
knitr::include_graphics(path = "figures/NEOF.png", auto_pdf = TRUE)
``` 

# Overview
```{r, fig.align = 'center',out.width= '20%', echo=FALSE }
knitr::include_graphics(path = "figures/well.png", auto_pdf = TRUE)
``` 

This practical session aims to introduce you to Long read Quality Control. The topics covered are:

- [Acquiring the workshop data](#data)
- [ONT quality control](#ont)
- [PacBio quality control](#pb)
- [Practice data](#practice)

<a rel="license" href="http://creativecommons.org/licenses/by-nc-sa/4.0/"><img alt="Creative Commons Licence" style="border-width:0" src="https://i.creativecommons.org/l/by-nc-sa/4.0/88x31.png" /></a><br />This work is licensed under a <a rel="license" href="http://creativecommons.org/licenses/by-nc-sa/4.0/">Creative Commons Attribution-NonCommercial-ShareAlike 4.0 International License</a>.

<!--chapter:end:01-Long_read_QC.Rmd-->

```{r include=FALSE, cache=FALSE}
suppressPackageStartupMessages({
  library(webexercises)
})

knitr::knit_hooks$set(webex.hide = function(before, options, envir) {
  if (before) {
    if (is.character(options$webex.hide)) {
      hide(options$webex.hide)
    } else {
      hide()
    }
  } else {
    unhide()
  }
})
```
# Data {#data}
```{r, fig.align = 'center',out.width= '20%', echo=FALSE }
knitr::include_graphics(path = "figures/Play.png", auto_pdf = TRUE)
``` 

## The data

The data for today is in the \"QC_workshop\" directory that you copied into your home directory (\"~\") in the Illumina QC day.

Check you have the directory.

```{bash, eval=FALSE}
ls ~
```

If you do not have the \"QC_workshop\" directory, the command to copy it all to your home directory is below.

```{bash eval=FALSE}
cp -r /pub39/tea/nsc006/NEOF/QC_workshop/ ~
```

<!--chapter:end:02-Intro.Rmd-->

```{r include=FALSE, cache=FALSE}
suppressPackageStartupMessages({
  library(webexercises)
})

knitr::knit_hooks$set(webex.hide = function(before, options, envir) {
  if (before) {
    if (is.character(options$webex.hide)) {
      hide(options$webex.hide)
    } else {
      hide()
    }
  } else {
    unhide()
  }
})
```
# (PART\*) ONT {.unnumbered}

# ONT background {#ont}
```{r, fig.align = 'center',out.width= '60%', echo=FALSE }
knitr::include_graphics(path = "figures/ONT_logo.png", auto_pdf = TRUE)
``` 

In this section we are going to carry out a quick QC of ONT data using the tool suite NanoPack and the tool Porechop. For this we will use the fastq files.

Before carrying out analysis we will cover some background of ONT data. This will include [file formats](#ont_file_formats), & [basecalling](#ont_basecalling).

## File formats {#ont_file_formats}
```{r, fig.align = 'center',out.width= '15%', echo=FALSE }
knitr::include_graphics(path = "figures/files.png", auto_pdf = TRUE)
``` 

Although we will only be using fastq files for this tutorial it is important to know the other file formats you may encounter when working with ONT data.

### Fast5
```{r, fig.align = 'center',out.width= '10%', echo=FALSE }
knitr::include_graphics(path = "figures/fast5.png", auto_pdf = TRUE)
``` 

The raw data from ONT machines comes as Fast5 (`.fast5`) files. This contains the signal data from the pore which can be processed into more useful files.

Fast5 files can contain:

+ Raw signal data
+ Run metadata
+ fastq-basecalls
+ Other additional analyses

The Fast5 file format is an implementation of the HDF5 file format specifically for ONT sequencing data. For more information on Fast5 and a tool to interface with these types of file please see:
https://github.com/nanoporetech/ont_fast5_api

### Summary file

The MinION and GridION output a single sequencing summary file in addition to the Fast5 files. This file contains metdata which describes each sequenced read.

We will not use Fast5 or summary files for this tutorial as they are not needed most of the time. It is likely these files will not be initially provided to you by a sequencing service as they are very large. However, if you do require them you can always ask but be careful with how long the sequencing centre will retain your data.

### BAM file
```{r, fig.align = 'center',out.width= '10%', echo=FALSE }
knitr::include_graphics(path = "figures/bam.png", auto_pdf = TRUE)
``` 

__BAM__ files (`.bam`) are the binary version of __SAM__ files (`.sam`).

- __SAM__ stands for "Sequence Alignment/Map" (the B in __BAM__ is Binary).
- __SAM__ files are tab delimited and contain alignment information. 
- __BAM__ files are not human readable whilst __SAM__ files are.
- __SAM__ files are larger than __BAM__.

Generally programs that can work on __SAM__ files can also work on __BAM__ files. Due to the size difference it is preferable to store data in __BAM__ format over __SAM__. Even though a BAM file is smaller than its matching __SAM__ file, __BAM__ files are still very large and can be >100GB.

It can be useful to contain unaligned reads from sequencing machines (e.g\. PacBio and ONT) in __BAM__ files as they can contain more metadata in the header and per-record auxiliary tags compared to __fastq__ files. 

If you are working with __SAM__/__BAM__ files the following link will prove useful: 
https://www.htslib.org/

For more information on the __SAM__ and __BAM__ format please see: 
https://samtools.github.io/hts-specs/SAMv1.pdf


### fastq file
```{r, fig.align = 'center',out.width= '10%', echo=FALSE }
knitr::include_graphics(path = "figures/fastq.png", auto_pdf = TRUE)
``` 

The __fastq__ file format is very consistent and so there is no real difference between __fastq__ files for Illumina, PacBio, and ONT data.

All __fastq__ files will contain a number of lines divisible by 4. This is because each entry/sequence will take up four lines consisting of the following information:

1. Header for __fastq__ entry known as the __fastq header__. 
    + This always begins with a __\'@\'__
    + This is where you might see the most difference between different data. Different machines and different public databases will use different formats for __fastq__ headers.
2. Sequence content
3. Quality header 
    + Always begins with a \'+\'. Sometimes also contains the same information as __fastq__ header.  
4. Quality
    + Each base in the 2nd line will have a corresponding quality value in this line.
    + Note this uses Phred encoding, of which there are different formats. Most, if not all, new files will use Phred+33 but be careful if you are using older data as it may use a different one. See the \"Encoding\" section in the following link for more info: https://en.wikipedia.org/wiki/FASTQ_format.
    +  NOTE: __\'@\'__ can be used as a quality value.

An example of the information of one sequence/entry is:

__\@Sequence 1__  
__CTGTTAAATACCGACTTGCGTCAGGTGCGTGAACAACTGGGCCGCTTT__  
__+__  
__=<<<=>\@\@\@ACDCBCDAC\@BAA\@BA\@BBCBBDA\@BB\@>CD\@A\@B?B\@\@__

## Basecalling {#ont_basecalling}
```{r, fig.align = 'center',out.width= '15%', echo=FALSE }
knitr::include_graphics(path = "figures/guppy.png", auto_pdf = TRUE)
``` 

In short, basecalling is the converting of ONT signals to bases and quality. This step converts the __Fast5__ files to __BAM__ and/or __fastq__ files.

There are many tools to carry out basecalling with ONT data. ONT sequencing machines carry this out with in built programs. However, if you are interested in the tools used for this the primary ones are Guppy and Albacore. Unfortunately it is quite hard to find information about these tools unless you own an ONT machine.

Basecalling with Guppy tutorial:

https://denbi-nanopore-training-course.readthedocs.io/en/latest/basecalling/basecalling.html

Basecalling with Albacore tutorial:

https://denbi-nanopore-training-course.readthedocs.io/en/stable/basecalling/basecalling.html

<!--chapter:end:03-ONT_background.Rmd-->

```{r include=FALSE, cache=FALSE}
suppressPackageStartupMessages({
  library(webexercises)
})

knitr::knit_hooks$set(webex.hide = function(before, options, envir) {
  if (before) {
    if (is.character(options$webex.hide)) {
      hide(options$webex.hide)
    } else {
      hide()
    }
  } else {
    unhide()
  }
})
```
# ONT Setup
```{r, fig.align = 'center',out.width= '10%', echo=FALSE }
knitr::include_graphics(path = "figures/conda.png", auto_pdf = TRUE)
```

We need to initialise the environment to use the programs we need. This needs to be done in each terminal you want to run the below commands in. Each terminal will only remember what you have told that terminal.

Run the following command in a new terminal. To open a new terminal window, right click on the main screen, choose `Applications` -> `Shell` -> `bash`.

```{bash eval=FALSE}
. usenanopack-1.1.0
```

Before carrying out any more commands we will move into the relevant directory.

```{bash eval=FALSE}
cd ~/QC_workshop/ONT_QC/
```

Look in the directory called data and you will notice there are a few directories. You can see the contents of all these directories with the below command.

```{bash eval=FALSE}
ls data/*
```

Each directory has one fastq file. ONT data likes to be organised with data for one sample being in one directory.

To start we will only use the fastq file within the directory called \"Acinetobacter\". As you may have figured out, this contains ONT sequencing data of an Acinetobacter genome. Specifically the data is a subset of the SRA (Sequence Read Archive) Run: SRR7119550.

<!--chapter:end:04-ONT_setup.Rmd-->

```{r include=FALSE, cache=FALSE}
suppressPackageStartupMessages({
  library(webexercises)
})

knitr::knit_hooks$set(webex.hide = function(before, options, envir) {
  if (before) {
    if (is.character(options$webex.hide)) {
      hide(options$webex.hide)
    } else {
      hide()
    }
  } else {
    unhide()
  }
})
```
# NanoStat
```{r, fig.align = 'center',out.width= '15%', echo=FALSE }
knitr::include_graphics(path = "figures/metrics.png", auto_pdf = TRUE)
``` 

The first step is to acquire stats for the sequences in our fastq file. We will use `NanoStat` (https://github.com/wdecoster/nanostat).

`NanoStat` is one of the many tools contained in the `NanoPack` suite (https://github.com/wdecoster/nanopack). We will also use the tools `NanoPlot` and `NanoFilt` downstream.

## NanoStat: run
```{r, fig.align = 'center',out.width= '15%', echo=FALSE }
knitr::include_graphics(path = "figures/enter_key_black.png", auto_pdf = TRUE)
```

We want to have a tidy set of directories at the end of this analysis. It would be an untidy mess if we had all the output files in the current directory. We will therefore be making directories and subdirectories to send our output.

```{bash eval=FALSE}
mkdir nanostats
```

Finally we will now run `NanoStat.` The options used are:

- `-n` : File name/path for the output.
- `-t` : Number of threads the script can use.
- `--fastq` : Input data is in fastq format. Other options that can be used are `--fasta`, `--summary`, and `--bam`.

```{bash eval=FALSE}
NanoStat -n nanostats/Acinetobacter_nanostats.tsv \
-t 4 --fastq ./data/Acinetobacter/Acinetobacter_genome_ONT.fastq
```

When the command has finished you can look at the output text file. In this case we will use the convenient `less` command.

```{bash eval=FALSE}
less nanostats/Acinetobacter_nanostats.tsv
```

## NanoStat: output {#nanostat_output}
```{r, fig.align = 'center',out.width= '15%', echo=FALSE }
knitr::include_graphics(path = "figures/output_file_black.png", auto_pdf = TRUE)
``` 

The file contains four sections with informative headers. These are:

1. __General summary__
   - A list of general summary metrics.
2. __Number, percentage and megabases of reads above quality cutoffs__
   - Based on the mean quality value of a read.
   - The \">Q5\" line shows the number and % of reads with a mean quality above Q5. It also shows the total amount of megabases these reads contain.
3. __Top 5 highest mean basecall quality scores and their read lengths__
   - Shows the top 5 reads with the highest mean quality scores.
4. __Top 5 longest reads and their mean basecall quality score__
   - Shows the top 5 longest reads.
 
## NanoStat: MCQs
```{r, fig.align = 'center',out.width= '15%', echo=FALSE }
knitr::include_graphics(path = "figures/question_bubble_black.png", auto_pdf = TRUE)
```

With the output of `NanoStat` and the above info on the output please answer the following Multiple-choice questions.

```{r, echo = FALSE}
opts_p <- c("__9.2__", "__10.2__", answer="__12.3__")
```
1.  What is the mean base quality for the read with the highest mean base quality?`r longmcq(opts_p)`

```{r, echo = FALSE}
opts_p <- c(answer="__9.2__", "__10.2__", "__12.3__")
```
2. What is the mean base quality for the longest read?`r longmcq(opts_p)`

```{r, echo = FALSE}
opts_p <- c("__9.2__", answer="__10.2__", "__12.3__")
```
3. What is the mean read quality for all the data?`r longmcq(opts_p)`

```{r, echo = FALSE}
opts_p <- c(answer="__5,770__", "__6,700.5__", "__25,072__")
```
4. What is the length of the read with the 2nd highest mean quality? `r longmcq(opts_p)`

```{r, echo = FALSE}
opts_p <- c("__5,770__", "__6,700.5__", answer="__25,072__")
```
5.  What is the length of the 3rd longest read? `r longmcq(opts_p)`

```{r, echo = FALSE}
opts_p <- c("__5,770__", answer="__6,700.5__", "__25,072__")
```
6.  What is the median read length of the data? `r longmcq(opts_p)`

```{r, echo = FALSE}
opts_p <- c("__0.2Mb__", "__8.8Mb__", answer="__14.1Mb__")
```
7.  To the nearest 0.1 Megabases (Mb, i.e. 100,000 bases), what is the total number of bases of the data? `r longmcq(opts_p)`

```{r, echo = FALSE}
opts_p <- c("__0.2Mb__", answer="__8.8Mb__", "__14.1Mb__")
```
8.  How many Megabases are within reads with a quality cutoff of >Q10?  `r longmcq(opts_p)`

```{r, echo = FALSE}
opts_p <- c("__0.2Mb__", "__8.8Mb__", answer="__14.1Mb__")
```
9. How many Megabases are within reads with a quality cutoff of >Q5? `r longmcq(opts_p)`

```{r, echo = FALSE}
opts_p <- c(answer="__0.2Mb__", "__8.8Mb__", "__14.1Mb__")
```
10.  How many Megabases are within reads with a quality cutoff of >Q12? `r longmcq(opts_p)`

## NanoStat: summary
```{r, fig.align = 'center',out.width= '15%', echo=FALSE }
knitr::include_graphics(path = "figures/sum_black.png", auto_pdf = TRUE)
```

We have carried out our first step of QC by quality checking the __fastq__ file with `NanoStat`. Overall the data has good quality, for ONT data >Q10 is good. However, there is some poor quality data so quality control will be needed.

<!--chapter:end:05-NanoStat.Rmd-->

```{r include=FALSE, cache=FALSE}
suppressPackageStartupMessages({
  library(webexercises)
})

knitr::knit_hooks$set(webex.hide = function(before, options, envir) {
  if (before) {
    if (is.character(options$webex.hide)) {
      hide(options$webex.hide)
    } else {
      hide()
    }
  } else {
    unhide()
  }
})
```
# Porechop
```{r, fig.align = 'center',out.width= '60%', echo=FALSE }
knitr::include_graphics(path = "figures/porechop_logo_knife.png", auto_pdf = TRUE)
``` 

`Porechop` is a tool to find and remove adapters from ONT data (https://github.com/rrwick/Porechop). Adapters are artificial sequences essential for sequencing but of no biological value and so you will typically want them removed.

`Porechop` is no longer supported but it still works and is a good tool. Depending on which basecaller was used on your data, adapter removal may have already been carried out. However, it is always best to run `porechop` if you are not sure. 

`Porechop` has a list of known adapters it will look for and remove. These contain:

- Ligation kit adapters
- Rapid kit adapters
- PCR kit adapters
- Barcodes
- Native barcoding
- Rapid barcoding

`Porechop` will look for these adapters at the start and end of each read. Then it will look for adapters within the sequence (known as middle adapters to `Porechop`). If it finds a middle adapter it will conclude that the read is chimeric (a recombinant read containing sequences from 2 or more reads) and split the read. Depending on the number of middle adapters the chimeric read may split into 2 or more reads.

## Porechop: run
```{r, fig.align = 'center',out.width= '15%', echo=FALSE }
knitr::include_graphics(path = "figures/enter_key_blue.png", auto_pdf = TRUE)
```

With all that explanation we will now run `Porechop.` Thankfully the command is relatively straight forward with the options:

- `-t`: Number of threads to be used.
- `-i`: Input path of a fasta file, fastq file, or a directory. If a directory is specified it will be recursively searched for fastq files.
- `-o`: Output path. This will either be a fastq or fasta file name.

```{bash eval=FALSE}
#Create output directory
mkdir porechop
#Run porechop command
porechop -t 4 -i ./data/Acinetobacter/Acinetobacter_genome_ONT.fastq \
-o porechop/Acinetobacter.porechop.fastq
```

`Porechop` will take a while to run. Whilst it is running, look at the screen output to get an idea of what it is doing. When finished, look at the bottom of the printed results.

## Porechop: MCQs
```{r, fig.align = 'center',out.width= '15%', echo=FALSE }
knitr::include_graphics(path = "figures/question_bubble_blue.png", auto_pdf = TRUE)
```

```{r, echo = FALSE}
opts_p <- c("__0__", answer="__1,952__", "__79,015__")
```
1. How many reads had adapters trimmed from their start? `r longmcq(opts_p)`

```{r, echo = FALSE}
opts_p <- c("__0__", "__1,952__", answer="__79,015__")
```
2.  How many bases were removed by adapters being trimmed from the end of reads? `r longmcq(opts_p)`

```{r, echo = FALSE}
opts_p <- c(answer="__0__", "__1,952__", "__79,015__")
```
3. How many reads were split based on middle adapters? `r longmcq(opts_p)`

## Porechop: summary
```{r, fig.align = 'center',out.width= '15%', echo=FALSE }
knitr::include_graphics(path = "figures/sum_blue.png", auto_pdf = TRUE)
```

`Porechop` did its job and removed adapters successfully. The next step is to view the quality of our reads prior to so we can be informed when we filter out poor quality reads and bases.

For more uses of `Porechop` please see the below links:

- https://github.com/rrwick/Porechop#quick-usage-examples
- https://github.com/rrwick/Porechop#full-usage

<!--chapter:end:06-Porechop.Rmd-->

```{r include=FALSE, cache=FALSE}
suppressPackageStartupMessages({
  library(webexercises)
})

knitr::knit_hooks$set(webex.hide = function(before, options, envir) {
  if (before) {
    if (is.character(options$webex.hide)) {
      hide(options$webex.hide)
    } else {
      hide()
    }
  } else {
    unhide()
  }
})
```
# NanoPlot
```{r, fig.align = 'center',out.width= '15%', echo=FALSE }
knitr::include_graphics(path = "figures/nanoplot_hexes.png", auto_pdf = TRUE)
``` 

`NanoPlot` can be thought of as the `fastqc` for ONT data. It produces a lot of useful visualisations to investigate the quality of ONT sequencing data. It can be used for __fastq__, __fasta__, __BAM__, and sequencing summary files. The link for its github page is: https://github.com/wdecoster/NanoPlot

ONT data has much lower quality scores than Illumina with Q10 being good. If you have enough coverage and length the low quality can be corrected by downstream processes not covered in this tutorial.

## NanoPlot: run
```{r, fig.align = 'center',out.width= '15%', echo=FALSE }
knitr::include_graphics(path = "figures/enter_key_green.png", auto_pdf = TRUE)
```

Prior to running `NanoPlot` we will make a directory for the `NanoPlot` output. As `NanoPlot` creates a lot of files, we'll make a subdirectory for the `NanoPlot` output for the porechopped data.

```{bash eval=FALSE}
mkdir nanoplot
mkdir nanoplot/porechop
```

Now to run `NanoPlot`. The options we will use are:

- `-t`: Number of threads to be used.
- `--fastq`: Specifies the input path which is a fastq file.
- `-o`: Directory where the output will be created.
- `-p`: Prefix of output files. It is useful to have "_" at the end of the prefix.
- `--plots`: Specifies what type of bivariate plot is to be created (more on this later). I find `hex` to be the best. 

```{bash eval=FALSE}
NanoPlot -t 4 \
--fastq porechop/Acinetobacter.porechop.fastq \
-o nanoplot/porechop -p Acinetobacter_ \
--plots hex 
```

You may get the below warning. This is fine and can be ignored.

```{bash, eval = FALSE}
/pub39/tea/nsc006/anaconda3/lib/python3.7/_collections_abc.py:702: MatplotlibDeprecationWarning: The global colormaps dictionary is no longer considered public API.
  return len(self._mapping)
/pub39/tea/nsc006/anaconda3/lib/python3.7/_collections_abc.py:720: MatplotlibDeprecationWarning: The global colormaps dictionary is no longer considered public API.
  yield from self._mapping
```

## NanoPlot: output
```{r, fig.align = 'center',out.width= '15%', echo=FALSE }
knitr::include_graphics(path = "figures/output_file_green.png", auto_pdf = TRUE)
``` 

List the files in the output directory.

```{bash, eval=FALSE}
ls nanoplot/porechop/
```

There are quite a few files. These should all start with \"Acinetobacter_\" thanks to the `-p` option. To quickly check all the results we can open the report html file with firefox.

```{bash eval=FALSE}
firefox nanoplot/porechop/Acinetobacter_NanoPlot-report.html
```

The first section contains [`NanoStat` output](#nanostat_output). Quickly look over this and see how it compares to the `NanoStat` output of the pre-porechopped reads.

`r hide("NanoStat of pre-porechopped reads")`
```{r, echo=FALSE}
#Create dataframe of the nanostat pre-porechopped data
df <- read.csv("files/Acinetobacter_nanostats.tsv", 
               sep = ":",
               check.names=FALSE)
df[,2] <- trimws(df[,2],which="both", whitespace="[ \t\r\n]")
#
colnames(df) <- c("General summary","values")
#Display table with flextable
#https://ardata-fr.github.io/flextable-book/index.html
library("flextable")
ft <- flextable(df)
ft <- autofit(ft)
ft
```
`r unhide()`

After the __Summary Statistics__ section there is a __Plots__ section. This contains the plots:

- __Histogram of read lengths__
   - Histogram of \"Number of _reads_\" (y) against \"Read length\" (x).
- __Histogram of read lengths after log transformations__
   - Histogram of \"Number of _reads_\" (y) against log transformed \"Read length\" (x).
- __Weighted Histogram of read lengths__
   - Histogram of \"Number of _bases_\" (y) against \"Read length\" (x).
- __Weighted Histogram of read lengths after log transformation__
   - Histogram of \"Number of _bases_\" (y) against log transformed \"Read length\" (x).
- __Dynamic histogram of Read length__
   - An interactive Histogram of \"Number of reads\" (y) against \"Read length\" (x) produced by [plotly](https://plotly.com/).
   - You can zoom into areas by clicking and dragging. 
      - Make boxes to zoom into a specific area. 
      - Click and drag left or right only to zoom into a specific part of the x axis. 
      - Click and drag up or down only to zoom into a specific part of the y axis. 
      - Click the home icon (`r icons::ionicons("home")`) on the top right to reset the axes.
      - Hover over a bar to see the specific values of it.
- __Yield by length__
   - Plot showing the \"Cumulative yield for minimal length\" (y) by \"Read length\" (x).
   - The cumulative yield is measured in Gigabases (billion bases).
   - This plot is useful to know how many bases you would retain if you filtered reads based on read length.
- __Read lengths vs Average read quality plot using hexagonal bins__
   - This is the most informative plot. It is the bivariate plot we indicated we wanted as hexagonal with the option `--plots hex`.
   - Each hex is a bin with a darker colour representing more reads.
   - At the top of the plot is a histogram of Number of reads against read lengths.
   - At the right of the plot is a sideways histogram of number of reads against average read quality.

## NanoPlot: MCQs
```{r, fig.align = 'center',out.width= '15%', echo=FALSE }
knitr::include_graphics(path = "figures/question_bubble_green.png", auto_pdf = TRUE)
```
Please attempt the below MCQs with the information in the report html.

```{r, echo = FALSE}
opts_p <- c(answer="__10.2__", "__6,962.1__", "__8.8Mb__")
```
1. What is the mean read quality for all the data?`r longmcq(opts_p)`

```{r, echo = FALSE}
opts_p <- c("__10.2__", answer="__6,962.1__", "__8.8Mb__")
```
2.  What is the mean read length of the data? `r longmcq(opts_p)`

```{r, echo = FALSE}
opts_p <- c("__10.2__", "__6,962.1__", answer="__8.8Mb__")
```
3.  How many Megabases are within reads with a quality cutoff of >Q10?  `r longmcq(opts_p)`

```{r, echo = FALSE}
opts_p <- c("__10__", "__146__", answer="__4Mb__")
```
4. Approximately how many bases would you retain if you filtered out/removed reads with a length <10,000bp (Yield by length)? `r longmcq(opts_p)`

```{r, echo = FALSE}
opts_p <- c("__10__", answer="__146__", "__4Mb__")
```
5. How many reads have a length between 6,000-6,500 (Dynamic histogram of read lengths) `r longmcq(opts_p)`

```{r, echo = FALSE}
opts_p <- c(answer="__10__", "__146__", "__4Mb__")
```
6. How many read have a length between 15K-15.499K (Dynamic histogram of read lengths)? `r longmcq(opts_p)`

```{r, echo = FALSE}
opts_p <- c("__6-8__", "__8-10__", answer="__10-12__")
```
7. Look at the "Read lengths vs Average read quality plot". In what average quality range are the majority of reads? `r longmcq(opts_p)`

## NanoPlot: summary
```{r, fig.align = 'center',out.width= '15%', echo=FALSE }
knitr::include_graphics(path = "figures/sum_green.png", auto_pdf = TRUE)
```

We have use `NanoPlot` to produce read quality and length stats in table and plot form. Our data still looks good after porechopping but we still need to carry out some filtering.

You can use `NanoPlot` with different input files. This will give you different plots. For more details please see the \"Plots Generated\" section on: https://github.com/wdecoster/NanoPlot.

<!--chapter:end:07-NanoPlot.Rmd-->

```{r include=FALSE, cache=FALSE}
suppressPackageStartupMessages({
  library(webexercises)
})

knitr::knit_hooks$set(webex.hide = function(before, options, envir) {
  if (before) {
    if (is.character(options$webex.hide)) {
      hide(options$webex.hide)
    } else {
      hide()
    }
  } else {
    unhide()
  }
})
```
# NanoFilt
```{r, fig.align = 'center',out.width= '10%', echo=FALSE }
knitr::include_graphics(path = "figures/quality_filtering.png", auto_pdf = TRUE)
```

`NanoFilt` can be used to remove/filter reads by quality and/or read length (https://github.com/wdecoster/nanofilt). This is very useful as you will most likely want long and good quality reads for downstream processes, such as genome assemblies.

It is always important to know what your data is and if your planned filtering and trimming is appropriate. For example, you may be working with amplicon data where the read lengths will vary between 500bp and 750bp. In that case it is useful to set a min length of 500 and a maxlength of 750.

This data is from a genome so wanting long and high quality reads is appropriate. Overall the quality of our data looks good. The main exception is the low amount of sequencing/coverage. The low coverage is because it is a subset of the whole data for workshop purposes (the full dataset would take too long to run here). There are some shorter length and lower quality reads which we will remove.

## NanoFilt: run
```{r, fig.align = 'center',out.width= '15%', echo=FALSE }
knitr::include_graphics(path = "figures/enter_key_purple.png", auto_pdf = TRUE)
```

First we will create an output directory.
```{bash eval=FALSE}
mkdir nanofilt
```

We will filter out sequences that are shorter than 500bp (`-l 500bp`) and filter out sequences with a average read quality less than Q10 (`-q 10`). We have chosen these values as they are a good default for genomic data. Q10 appeared to be a good choice from NanoPlot as the majority of reads had a quality of >Q10.

```{bash eval=FALSE}
cat porechop/Acinetobacter.porechop.fastq | \
NanoFilt -l 500 -q 10 > \
nanofilt/Acinetobacter.nanofiltq10minlen500.porechop.fastq
```

Are the chosen options appropriate? Let\'s find out.

<!--chapter:end:08-NanoFilt.Rmd-->

```{r include=FALSE, cache=FALSE}
suppressPackageStartupMessages({
  library(webexercises)
})

knitr::knit_hooks$set(webex.hide = function(before, options, envir) {
  if (before) {
    if (is.character(options$webex.hide)) {
      hide(options$webex.hide)
    } else {
      hide()
    }
  } else {
    unhide()
  }
})
```
# Final check
```{r, fig.align = 'center',out.width= '15%', echo=FALSE }
knitr::include_graphics(path = "figures/finish.png", auto_pdf = TRUE)
```

After trimming or filtering reads (quality control) it is always important to carry out a quality check. We will therefore run NanoPlot again.

__Note__: We are using a long informative output directory name. This is important as we may need to rerun NanoFilt a few more times with different options until we are happy. 

```{bash eval=FALSE}
#Make an output directory before running NanoPlot
mkdir nanoplot/nanofiltq10minlen500_porechop
#Run NanoPlot on the fastq file with the filtered and porechopped data
NanoPlot -t 4 \
--fastq nanofilt/Acinetobacter.nanofiltq10minlen500.porechop.fastq \
-o nanoplot/nanofiltq10minlen500_porechop/ -p Acinetobacter_ \
--plots hex 
```

Now inspect the report

```{bash, eval=FALSE}
firefox nanoplot/nanofiltq10minlen500_porechop/Acinetobacter_NanoPlot-report.html
```

Ultimately when quality checking we need to take into account how much data we are left with. We need to make sure we have a good amount of bases and reads for our application. However, we also want to make sure we aren\'t left with a lot of poor quality data.

Depending on our application the quality of the reads may be less important, or the other way around. This is the same with the amount of reads and bases. What is needed will be clearer when you know what type of data you have. These considerations will be covered in our future workshops of the specific data types

A quick example is data for a genome assembly. For this type of analysis we will want our number of bases to reach a certain coverage. We\'ll want a decent quality but a high coverage can help overcome lower quality (more on this in a future workshop). Generally for ONT assemblies you will want 20X - 100X coverage with a higher coverage producing a better assembly (100x coverage= 100bp sequencing data per 1bp of the genome size). Higher coverages may be worse as it can be too much information for an assembly algorithm to cope with.

In addition to the amount of bases, the length of reads is also important. One read that is 10kbp long is most likely better than 10 reads that are each 1kbp long for a genome assembly. This is because less reads will need to be assembled.

Keeping all this in mind, let\'s say this sequencing is for a genome with a size of 0.5 Mbp (0.5 million base pairs). We want at least 20X coverage (i.e 10,000,000 total bases). Do we have this for the data that was porechopped and nanofiltered?

If not, try running NanoFilt with a different choices for `--q` until you do. 

+ What is an option that works to get the desired coverage? 
+ Is the overall quality decent (Mean & Median read quality > 10)?
+ What is the N50 of your QC\'d reads?
+ How many extra reads and bases did you retain compared to the Q10 filtering?

Tips:

+ Make sure your new output paths have unique names.
+ Use NanoPlot results to compare the Q10 filtering to your own. The NanoStat part is especially helpful.

<!--chapter:end:09-ONT_final_check.Rmd-->

