<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 5 Practice | Long read Quality Control</title>
  <meta name="description" content="NEOF book for the Long read quality control workshop" />
  <meta name="generator" content="bookdown 0.24 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 5 Practice | Long read Quality Control" />
  <meta property="og:type" content="book" />
  
  <meta property="og:image" content="/figures/NEOF.png" />
  <meta property="og:description" content="NEOF book for the Long read quality control workshop" />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 5 Practice | Long read Quality Control" />
  
  <meta name="twitter:description" content="NEOF book for the Long read quality control workshop" />
  <meta name="twitter:image" content="/figures/NEOF.png" />

<meta name="author" content="Matthew Gemmell &amp; Helen Hipperson" />


<meta name="date" content="2021-12-02" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  <link rel="shortcut icon" href="figures/NEOF_favicon.png" type="image/x-icon" />
<link rel="prev" href="04-PB.html"/>
<link rel="next" href="06-Furter_resources.html"/>
<script src="libs/header-attrs-2.11/header-attrs.js"></script>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.0.1/anchor-sections.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.0.1/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li class="toc-logo"><a href="https://neof.org.uk/" target="blank"><img src="figures/neof_small_logo.png"></a></li>
<li><a>Long read QC</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="01-Long_read_QC.html"><a href="01-Long_read_QC.html"><i class="fa fa-check"></i><b>1</b> Overview</a></li>
<li class="chapter" data-level="2" data-path="02-Intro.html"><a href="02-Intro.html"><i class="fa fa-check"></i><b>2</b> Data</a>
<ul>
<li class="chapter" data-level="2.1" data-path="02-Intro.html"><a href="02-Intro.html#the-data"><i class="fa fa-check"></i><b>2.1</b> The data</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="03-ONT.html"><a href="03-ONT.html"><i class="fa fa-check"></i><b>3</b> ONT</a>
<ul>
<li class="chapter" data-level="3.1" data-path="03-ONT.html"><a href="03-ONT.html#file-formats"><i class="fa fa-check"></i><b>3.1</b> File formats</a>
<ul>
<li class="chapter" data-level="3.1.1" data-path="03-ONT.html"><a href="03-ONT.html#fast5"><i class="fa fa-check"></i><b>3.1.1</b> Fast5</a></li>
<li class="chapter" data-level="3.1.2" data-path="03-ONT.html"><a href="03-ONT.html#summary-file"><i class="fa fa-check"></i><b>3.1.2</b> Summary file</a></li>
<li class="chapter" data-level="3.1.3" data-path="03-ONT.html"><a href="03-ONT.html#bam-file"><i class="fa fa-check"></i><b>3.1.3</b> BAM file</a></li>
<li class="chapter" data-level="3.1.4" data-path="03-ONT.html"><a href="03-ONT.html#fastq-file"><i class="fa fa-check"></i><b>3.1.4</b> fastq file</a></li>
</ul></li>
<li class="chapter" data-level="3.2" data-path="03-ONT.html"><a href="03-ONT.html#basecalling"><i class="fa fa-check"></i><b>3.2</b> Basecalling</a></li>
<li class="chapter" data-level="3.3" data-path="03-ONT.html"><a href="03-ONT.html#qc-of-fastq-file"><i class="fa fa-check"></i><b>3.3</b> QC of fastq file</a>
<ul>
<li class="chapter" data-level="3.3.1" data-path="03-ONT.html"><a href="03-ONT.html#nanostat"><i class="fa fa-check"></i><b>3.3.1</b> NanoStat</a></li>
<li class="chapter" data-level="3.3.2" data-path="03-ONT.html"><a href="03-ONT.html#porechop"><i class="fa fa-check"></i><b>3.3.2</b> Porechop</a></li>
<li class="chapter" data-level="3.3.3" data-path="03-ONT.html"><a href="03-ONT.html#nanoplot"><i class="fa fa-check"></i><b>3.3.3</b> NanoPlot</a></li>
<li class="chapter" data-level="3.3.4" data-path="03-ONT.html"><a href="03-ONT.html#nanofilt"><i class="fa fa-check"></i><b>3.3.4</b> NanoFilt</a></li>
<li class="chapter" data-level="3.3.5" data-path="03-ONT.html"><a href="03-ONT.html#final-check"><i class="fa fa-check"></i><b>3.3.5</b> Final check</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="4" data-path="04-PB.html"><a href="04-PB.html"><i class="fa fa-check"></i><b>4</b> PacBio</a>
<ul>
<li class="chapter" data-level="4.1" data-path="04-PB.html"><a href="04-PB.html#file-formats-1"><i class="fa fa-check"></i><b>4.1</b> File formats</a></li>
<li class="chapter" data-level="4.2" data-path="04-PB.html"><a href="04-PB.html#continuous-long-reads"><i class="fa fa-check"></i><b>4.2</b> Continuous Long Reads</a>
<ul>
<li class="chapter" data-level="4.2.1" data-path="04-PB.html"><a href="04-PB.html#quality-check"><i class="fa fa-check"></i><b>4.2.1</b> Quality check</a></li>
<li class="chapter" data-level="4.2.2" data-path="04-PB.html"><a href="04-PB.html#filtering"><i class="fa fa-check"></i><b>4.2.2</b> Filtering</a></li>
</ul></li>
<li class="chapter" data-level="4.3" data-path="04-PB.html"><a href="04-PB.html#circular-consensus-sequences"><i class="fa fa-check"></i><b>4.3</b> Circular Consensus Sequences</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="05-Practice.html"><a href="05-Practice.html"><i class="fa fa-check"></i><b>5</b> Practice</a>
<ul>
<li class="chapter" data-level="5.1" data-path="05-Practice.html"><a href="05-Practice.html#ont-data"><i class="fa fa-check"></i><b>5.1</b> ONT data</a></li>
<li class="chapter" data-level="5.2" data-path="05-Practice.html"><a href="05-Practice.html#pacbio-data"><i class="fa fa-check"></i><b>5.2</b> PacBio data</a></li>
</ul></li>
<li class="appendix"><span><b>Appendix</b></span></li>
<li class="chapter" data-level="A" data-path="06-Furter_resources.html"><a href="06-Furter_resources.html"><i class="fa fa-check"></i><b>A</b> Videos</a></li>
<li class="chapter" data-level="B" data-path="06-Furter_resources.html"><a href="06-Furter_resources.html#long-read-sequencing-resources"><i class="fa fa-check"></i><b>B</b> Long read sequencing resources</a></li>
<li class="chapter" data-level="C" data-path="06-Furter_resources.html"><a href="06-Furter_resources.html#pacbio-resources"><i class="fa fa-check"></i><b>C</b> PacBio resources</a></li>
<li class="chapter" data-level="D" data-path="06-Furter_resources.html"><a href="06-Furter_resources.html#ont-resources"><i class="fa fa-check"></i><b>D</b> ONT resources</a></li>
<li class="chapter" data-level="E" data-path="06-Furter_resources.html"><a href="06-Furter_resources.html#tools"><i class="fa fa-check"></i><b>E</b> Tools</a></li>
<li class="chapter" data-level="F" data-path="06-Furter_resources.html"><a href="06-Furter_resources.html#misc"><i class="fa fa-check"></i><b>F</b> Misc</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Long read Quality Control</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="practice" class="section level1" number="5">
<h1><span class="header-section-number">Chapter 5</span> Practice</h1>
<p><img src="figures/weight.png" width="20%" style="display: block; margin: auto;" /></p>
<p>This section contains information on more data that you can carry out QC (Quality control and checking) as self guided practice.</p>
<div id="ont-data" class="section level2" number="5.1">
<h2><span class="header-section-number">5.1</span> ONT data</h2>
<p><img src="figures/ONT_logo.png" width="60%" style="display: block; margin: auto;" />
The extra data consists of 2 subsetted runs from Bio Project: PRJNA477342 (<a href="https://www.ncbi.nlm.nih.gov/bioproject/?term=prjna477342" class="uri">https://www.ncbi.nlm.nih.gov/bioproject/?term=prjna477342</a>).</p>
<p>The data sets are:</p>
<ul>
<li>Flavobacterium
<ul>
<li>Genomic data for a <strong>Flavobacterium</strong> isolate.</li>
<li>Original data from: SRA run SRR7449788.</li>
</ul></li>
<li>Pseudonocardia
<ul>
<li>Genomic data for a <strong>Pseudonocardia</strong> isolate.</li>
<li>Original data from: SRA run SRR7447112.</li>
</ul></li>
</ul>
<p>These files are in the directory <code>~/QC_workshop/ONT_QC/data</code></p>
<p>Run through all the ONT QC steps for these two datasets. Use all the same parameters and options, making sure to change input and output file paths and prefixes where appropriate.</p>
<p><strong>Tips</strong>:</p>
<ul>
<li>If you are editing commands you previously ran it is best to change the outputs first so you don't accidentally overwrite previous results.</li>
<li>You can run the command <code>firefox nanoplot/*/*NanoPlot*html</code>. This will open all the NanoPlot report htmls in one firefox window. There will be a separate tab for each report.</li>
</ul>
<p><strong>Questions</strong>:</p>
<ol style="list-style-type: decimal">
<li>Which of the three datasets (Acinetobacter, Flavobacterium, and Pseudocardia) has the highest initial (pre porechopping) mean read quality (this may be more than one)?</li>
<li>Which of the three datasets has the highest initial mean read length?</li>
<li>Which of the three datasets has the longest read in the initial data? How long is the read and what is it mean basecall quality score?</li>
<li>Answer questions 1-3 for the porechopped data. Are the results of NanoStats significantly changed for any of the samples?</li>
<li>Look at the NanoPlot results for the porechopped and NanoFiltered data using <code>-l 500</code> and <code>-q 10</code>.
<ul>
<li>How many bases and reads are removed from the samples due to NanoFilt?</li>
<li>Are the parameters too stringent for some of the samples (losing too much data)?</li>
<li>Try out some different parameters until you are happy with the results.</li>
</ul></li>
</ol>
<p>Ultimately, knowing what is good and poor data takes experience and practice. The <code>-l 500</code> and <code>-q10</code> is a good default for ONT genomic data and may be good for other uses. As with all defaults you may need to change these for your particular data.</p>
<p>Hopefully that has been good practice and you are ready to QC your own data.</p>
</div>
<div id="pacbio-data" class="section level2" number="5.2">
<h2><span class="header-section-number">5.2</span> PacBio data</h2>
<p><img src="figures/Pacbio_logo.png" width="40%" style="display: block; margin: auto;" /></p>
<p>The extra data sets are from whole genome sequencing projects for three microbial isolates. These data are a subset of a multiplexed microbial sequencing run of 15 microbial isolates, available here: <a href="https://github.com/PacificBiosciences/DevNet/wiki/Microbial-Multiplexing-Data-Set---48-plex:-PacBio-Sequel-II-System,-Chemistry-v2.0,-SMRT-Link-v8.0-Analysis" class="uri">https://github.com/PacificBiosciences/DevNet/wiki/Microbial-Multiplexing-Data-Set---48-plex:-PacBio-Sequel-II-System,-Chemistry-v2.0,-SMRT-Link-v8.0-Analysis</a></p>
<p>The data sets are:</p>
<ul>
<li><em>Helicobacter pylori</em> in the file "bc1018.subreads.bam"</li>
<li><em>Methanocorpusculum labreanum</em> in the file "bc1037.subreads.bam"</li>
<li><em>Neisseria meningitidis</em> in the file "bc1044.subreads.bam"</li>
</ul>
<p>These files are in the directory <code>~/QC_workshop/PB_QC/data/PB_exercise</code></p>
<p>Run through the SequelTools quality check on these three data sets to generate the summary values and read length plots (there is no need to do the filtering step).</p>
<p><strong>Questions</strong>:</p>
<ol style="list-style-type: decimal">
<li>Which of the three samples contains the most subread data (Mb)?</li>
<li>Which samples has the highest subread N50? And which the lowest?</li>
<li>Which sample contains the longest subread, and what is it's approximate length?</li>
<li>Considering all three samples, do any of them look to have failed or sequenced very badly compared to the others?</li>
</ol>
<p>The filtering tool from SequelTools would likely be of limited use here as the longest subread and total subread lengths are very similar, as we would want in a whole genome CLR project. For PacBio data it is common for downstream tools (such as assemblers, or ccs for amplicon data) to apply some filtering and they require the raw data as input. We will cover specific examples in our future workshops that use PacBio sequences.</p>
<p>Hopefully that has been good practice and you are ready to QC your own data.</p>

</div>
</div>



            </section>

          </div>
        </div>
      </div>
<a href="04-PB.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="06-Furter_resources.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": null,
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": null,
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": null,
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "section"
}
});
});
</script>

</body>

</html>
